---
title: "Summary report of MR analyses for childhood BMI,  mediator: `r params$mediator_name`, and breast cancer `r params$breastcancer_dataset`"
author: "Marina Vabistsevits"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
 html_document:
    theme: lumen
    number_section: true
    toc: true
    toc_depth: 2
    toc_float:
      collapse: true
      smooth_scroll: true
params:
  mediator_name: mediator_name
  breastcancer_dataset: breastcancer_dataset

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, comment="")
library(readr)
library(data.table)
library(vroom)

library(tidyr)
library(purrr)
library(tibble)
library(dplyr)

library(TwoSampleMR)
library(MVMR)

library(ggplot2)
library(cowplot)
library(wesanderson)
library(kableExtra)

source("functions_mvmr.R")
source("functions.R")
```

### Set environment and load the data {-#set-environment-and-load-the-data} 

_Running the report with parameters:_

  * **Mediator**: `r params$mediator_name`
  * **Breast cancer output**: `r params$breastcancer_dataset`


```{r, warning=FALSE, include=FALSE}
# set path for pre-calculated data, outside the code repo
# `local` / `remote` (reading data from RDSF)
currently_working_env = "local"
source("set_paths.R")
set_paths(currently_working_env)

# metadata file
data_lookup<-read_csv(paste0("metadata/data_lookup.csv"))
```


```{r, warning=FALSE, include=TRUE, cache=TRUE, message=FALSE, comment=""}
# Load BMI exposures
early_bmi_file <- data_lookup %>% filter(trait == "Childhood BMI") %>% pull(tophits_data)
early_bmi_exp <- read_tsv(paste0(data_path_tophits, early_bmi_file))
dim(early_bmi_exp) # 115 in new

adult_bmi_file <- data_lookup %>% filter(trait == "Adult BMI") %>% pull(tophits_data)
adult_bmi_exp <- read_tsv(paste0(data_path_tophits, adult_bmi_file))
dim(adult_bmi_exp) # 173 in new

# Load BMI outcomes
early_bmi_gwas_file <- data_lookup %>% filter(trait == "Childhood BMI") %>% pull(full_data)
early_bmi_gwas <- vroom(paste0(data_path_gwas, early_bmi_gwas_file))
dim(early_bmi_gwas) 

adult_bmi_gwas_file <- data_lookup %>% filter(trait == "Adult BMI") %>% pull(full_data)
adult_bmi_gwas <- vroom(paste0(data_path_gwas, adult_bmi_gwas_file))
dim(adult_bmi_gwas) 

```

```{r, warning=FALSE, include=T, cache=T, message=FALSE, comment=""}
# Load mediator instrumets and full data / extract from MRBase
mediator_name <- params$mediator_name
data_lookup <- data_lookup %>% filter(trait==mediator_name)

# get current trait category (to be used for out file location)
trait_cat <- data_lookup %>% pull(trait_category) %>% unique()
results_path_sub <-  paste0(results_path, trait_cat,  "/") 

# get instruments
if (data_lookup$format == "tsv"){
  exposure_mediator <-fread(paste0(data_path_tophits, data_lookup$tophits_data))
  exposure_mediator$exposure <- mediator_name 
  exposure_mediator$data_source.exposure <- "textfile"
  
} else if (data_lookup$format == "mrbase") {
  exposure_mediator <- extract_instruments(data_lookup$mrbase.id) 
  if (is.null(exposure_mediator)){ exposure_mediator <- extract_instruments(data_lookup$mrbase.id, p1 = 10e-07)} # if no SNPs returned, try lower pval 
  exposure_mediator <- clump_data(exposure_mediator)
  exposure_mediator$exposure <- mediator_name
}

# Load mediator in outcome fro  MR-base or from textfile (full GWAS file) 
if (data_lookup$format == "mrbase"){
  mediator_early<- extract_outcome_data(
      snps = early_bmi_exp$SNP,
      outcomes = data_lookup$mrbase.id)
  mediator_early$outcome <- mediator_name
  
  mediator_adult<- extract_outcome_data(
      snps = adult_bmi_exp$SNP,
      outcomes = data_lookup$mrbase.id)
  mediator_adult$outcome <- mediator_name
} else {
  current_file <- paste0(data_path_gwas, data_lookup$full_data)
  mediator <- vroom(current_file) 

  mediator_early <- mediator %>% filter(SNP %in% early_bmi_exp$SNP)
  mediator_adult <- mediator %>% filter(SNP %in% adult_bmi_exp$SNP)
}
```


# Part. BMI (childhood/adult) as exposure, `r params$mediator_name` as outcome 

## Univariate MR analysis

### Harmonise

```{r, warning=FALSE}        
dat_early <- harmonise_data(exposure_dat = early_bmi_exp, 
                        outcome_dat = mediator_early)
dat_adult <- harmonise_data(exposure_dat = adult_bmi_exp, 
                        outcome_dat = mediator_adult)
```


### Perform MR

```{r, warning=FALSE}
res_early <- mr(dat_early,
                method_list=c('mr_ivw','mr_egger_regression','mr_weighted_median', 'mr_wald_ratio')) 
res_adult <- mr(dat_adult,
                method_list=c('mr_ivw','mr_egger_regression','mr_weighted_median', 'mr_wald_ratio')) 
```


### Tidy up results

- Split outcome names
- Generate odds ratios with 95% confidence intervals

```{r, warning=FALSE}
res_early_tidy<- res_early %>%
              split_outcome() %>% 
              split_exposure() %>% 
              generate_odds_ratios()

res_adult_tidy<- res_adult %>%
              split_outcome() %>% 
              split_exposure() %>%
              generate_odds_ratios()

out <- rbind(res_early_tidy, res_adult_tidy ) %>% 
          arrange(method) %>% 
          select(-c("id.outcome", "id.exposure"))

```

```{r, include=F}
report_dir <- paste0(results_path_sub, data_lookup$trait_file_name, "/", "report_",Sys.Date(), "_", params$breastcancer_dataset)
report_dir_w_backslash <- gsub(" ", "\\ ", report_dir, fixed=T) # create path vector escaping spaces, otherwise sytem call cant process it
if(!dir.exists(report_dir)){ system(paste("mkdir -p", report_dir_w_backslash))}
write_tsv(out, paste0(report_dir, "/MR_BMI-to-", data_lookup$trait_file_name ,".tsv")) # write to path with spaces
print(out)
```



```{r warning=FALSE, include=F}

pal <- wes_palette("FantasticFox1")[c(3,4)]
plot<-ggplot(out, aes(y=method, x=b, label=outcome, colour=exposure, shape=method)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=.3) +
  geom_point(size=2)+
  scale_color_manual(values=pal)+
  scale_shape_manual(values = c(19,20,8)) +
  geom_vline(xintercept=0, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~exposure, ncol=1)+
  labs(color = "",y = "", x = "Effect size",
       title=paste0("           Univariate MR results of BMI (exposures) and ", mediator_name, " (outcome)"))+
  theme(legend.position = "none",
        plot.title.position = "plot")
```

```{r plot, fig.height = 3, fig.width = 5, echo=F}
plot
```

## Sensitivity analyses

### Horizonal pleiotropy

The intercept term in MR Egger regression can be a useful indication of whether directional horizontal pleiotropy is driving the results of an MR analysis. To test for an average pleiotropic effect, we can assess the extent to which the MR-Egger intercept is non-zero (i.e. non-zero intercept is indicative of directional pleiotropy).

If `egger_intercept` is reasonably close to zero (e.g. 0.002), and p-value is high, there is no evidence of substantial directional pleitropic effect. 

```{r}
mr_pleiotropy_test(dat_early)
```

```{r}
mr_pleiotropy_test(dat_adult)
```


### Heterogenity

We can assess heterogeneity (difference in individual ratio estimates) across the set of genetic variants used. To do so, we need to obtain Q statistics for heterogeneity with respect to IVW and MR-Egger. 

When the Q statistic (Cochranâ€™s Q) is quite high and p-value is very small, there is evidence for heterogeneity. However, this heterogeneity does not translate into mean average directional effect if the `egger_intercept` is zero (or close). Therefore, such high Q-statistic can be an indicatorof one or more outliers in the analysis, which could indicate that variants that serve as outliers, are also violating the MR assumptions.

If Q > K-1, the estimates exhibit over-dispersion = heterogeniety; (K is the number of instruments)

If Q ~ K-1, no heterogeniety is detected

```{r, warning=FALSE, echo=F}
print(paste0("Number of instruments in Early BMI: ", unique(res_early_tidy$nsnp) ))
```

```{r}
mr_heterogeneity(dat_early, method_list=c("mr_egger_regression", "mr_ivw"))
```

```{r warning=FALSE, echo=F}
print(paste0("Number of instruments in Adult BMI: ", unique(res_adult_tidy$nsnp) ))
```

```{r}
mr_heterogeneity(dat_adult, method_list=c("mr_egger_regression", "mr_ivw")) 
```

### Scatter plot

Scatter	plots	show	BYG	plotted	against	BXG,	with	the	expectation	for	a	linear	dose-response	relationship	and	an	intercept	passing	through	zero.	Deviation	from this	may	be	indicative	of	heterogeneity	and	violations	of	MR	assumptions,	e.g.	due	to	horizontal	pleiotropy	(the	intercept	from	MR-Egger	is	a	formal	statistical	test	for	this).

```{r, warning=FALSE, echo=FALSE, message=FALSE}
mr_scatter_plot(res_early, dat_early)
mr_scatter_plot(res_adult, dat_adult)
```

### Single SNP MR analysis

```{r, warning=FALSE}
res_single_early <- mr_singlesnp(dat_early)
res_single_adult <- mr_singlesnp(dat_adult)
```

### Forest plot

Forest	plots	summarise	the	effect	of	the	exposure	on	the	outcome	due	to each	SNP	separately	as well	as	combined	across	all	SNPs	into	a	single	instrument.	These	 plots	help	to	visualise	heterogeneity	between	SNPs,	which	can	be	a	useful	indicator	of	 violations	of	MR	assumptions	(e.g.	due	to	horizontal	pleiotropy).	

```{r, warning=FALSE}
mr_forest_plot(res_single_early)
mr_forest_plot(res_single_adult)
```

### Funnel plot

Funnel	plots	display	the	 individual	Wald	ratios	for	each	SNP	plotted	against	their	precision,	where	asymmetry	is	 indicative	of	directional	horizontal	pleiotropy	(an	important	violation	of	MR	assumptions).	

Less precise causal estimates 'funnel' in from either side towards more precise estimates. When there is no pleiotropy, the funnel should be symmetric. If it's symmetric but points spread out, this indicates heterogeneity (and balanced pleiotropy). If funnel is asymmetic, there is directional pleiotropy. 

```{r, warning=FALSE}
mr_funnel_plot(res_single_early)
mr_funnel_plot(res_single_adult)
```

### Leave-one-out analysis

A	plot	of	a	leave-one-out	analysis	shows	the	influence	of	individual	SNPs	on	the	overall	effect	estimate,	which	can	beused	to	identifyimportant	outliers.	

```{r, warning=FALSE}
res_loo_early <- mr_leaveoneout(dat_early)
res_loo_adult <- mr_leaveoneout(dat_adult)
```

```{r, warning=FALSE, message=FALSE}
mr_leaveoneout_plot(res_loo_early)
mr_leaveoneout_plot(res_loo_adult)
```


\pagebreak


# Part. `r params$mediator_name` as exposure, Breast cancer as outcome 

## Univariate MR analysis

### Get outcome from MR-Base
 
```{r}
# exposure mediator
dim(exposure_mediator)

outcome_BC <- extract_outcome_data(
                snps = exposure_mediator$SNP,
                outcome = params$breastcancer_dataset,
                proxies = TRUE,
                rsq = 0.8, maf_threshold = 0.3) 

dim(outcome_BC) 
```
### Harmonise

```{r}
dat<- harmonise_data(exposure_dat = exposure_mediator, 
                      outcome_dat = outcome_BC)
```
### Perform MR

```{r}
res <- mr(dat, method_list=c('mr_ivw','mr_egger_regression','mr_weighted_median', 'mr_wald_ratio')) 
```

### Tidy up results

- Split outcome names
- Generate odds ratios with 95% confidence intervals

```{r, warning=FALSE}
res_tidy <- res %>%
            split_outcome() %>% 
            split_exposure() %>% 
            separate(outcome, "outcome", sep="[(]") %>% 
            select(-c("id.exposure")) %>% 
            generate_odds_ratios()

```

```{r, include=F}
report_dir <- paste0(results_path_sub, data_lookup$trait_file_name, "/", "report_",Sys.Date())
report_dir_w_backslash <- gsub(" ", "\\ ", report_dir, fixed=T) # create path vector escaping spaces, otherwise sytem call cant process it
if(!dir.exists(report_dir)){ system(paste("mkdir -p", report_dir_w_backslash))}
write_tsv(res_tidy, paste0(report_dir, "/MR_", data_lookup$trait_file_name ,"-to-BC.tsv"))
print(res_tidy)
```


```{r warning=FALSE, include=F}
pal <- wes_palette("Rushmore1")[c(3)]
plot2<-ggplot(res_tidy, aes(y=method, x=or, label=outcome, colour=exposure, shape=method)) +
  geom_errorbarh(aes(xmin=or_lci95, xmax=or_uci95), height=.3) +
  geom_point(size=2)+
  scale_color_manual(values=pal)+
  scale_shape_manual(values = c(19,20,8)) +
  geom_vline(xintercept=1, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~exposure, ncol=1)+
  labs(color = "",y = "", x = "Odds ratio",
       title=paste0(" Univariate MR results of ", mediator_name, " (exposure) and Breast Cancer (outcome)"))+
  theme(legend.position = "none",
        plot.title.position = "plot")
```

```{r plot2, fig.height = 2, fig.width = 5, echo=F}
plot2
```


## Sensitivity analyses


### Horizonal pleiotropy

The intercept term in MR Egger regression can be a useful indication of whether directional horizontal pleiotropy is driving the results of an MR analysis. To test for an average pleiotropic effect, we can assess the extent to which the MR-Egger intercept is non-zero (i.e. non-zero intercept is indicative of directional pleiotropy).

If `egger_intercept` is reasonably close to zero (e.g. 0.002), and p-value is high, there is no evidence of substantial directional pleitropic effect. 

```{r}
mr_pleiotropy_test(dat)
```


### Heterogenity

We can assess heterogeneity (difference in individual ratio estimates) across the set of genetic variants used. To do so, we need to obtain Q statistics for heterogeneity with respect to IVW and MR-Egger. 

When the Q statistic (Cochranâ€™s Q) is quite high and p-value is very small, there is evidence for heterogeneity. However, this heterogeneity does not translate into mean average directional effect if the `egger_intercept` is zero (or close). Therefore, such high Q-statistic can be an indicatorof one or more outliers in the analysis, which could indicate that variants that serve as outliers, are also violating the MR assumptions.

If Q > K-1, the estimates exhibit over-dispersion = heterogeniety; (K is the number of instruments)

If Q ~ K-1, no heterogeniety is detected

```{r, warning=FALSE, echo=F}
print(paste0("Number of instruments in ", params$mediator_name, ":", unique(res_tidy$nsnp)  ))
```

```{r, warning=FALSE}
mr_heterogeneity(dat) 
```

### Scatter plot

Scatter	plots	show	BYG	plotted	against	BXG,	with	the	expectation	for	a	linear	dose-response	relationship	and	an	intercept	passing	through	zero.	Deviation	from	this	may	be	indicative	of	heterogeneity	and	violations	of	MR	assumptions,	e.g.	due	to	horizontal	pleiotropy	(the	intercept	from	MR-Egger	is	a	formal	statistical	test	for	this).

```{r, warning=FALSE, echo=FALSE, message=FALSE}
mr_scatter_plot(res, dat)
```

### Single SNP MR analysis

```{r, warning=FALSE}
res_single <- mr_singlesnp(dat)
```

### Forest plot

Forest	plots	summarise	the	effect	of	the	exposure	on	the	outcome	due	to each	SNP	separately	as well	as	combined	across	all	SNPs	into	a	single	instrument.	These	 plots	help	to	visualise	heterogeneity	between	SNPs,	which	can	be	a	useful	indicator	of	 violations	of	MR	assumptions	(e.g.	due	to	horizontal	pleiotropy).	

```{r, warning=FALSE}
mr_forest_plot(res_single)
```

### Funnel plot

Funnel	plots	display	the	 individual	Wald	ratios	for	each	SNP	plotted	against	their	precision,	where	asymmetry	is	 indicative	of	directional	horizontal	pleiotropy	(an	important	violation	of	MR	assumptions).	

Less precise causal estimates 'funnel' in from either side towards more precise estimates. When there is no pleiotropy, the funnel should be symmetric. If it's symmetric but points spread out, this indicates heterogeneity (and balanced pleiotropy). If funnel is asymmetic, there is directional pleiotropy. 

```{r, warning=FALSE}
mr_funnel_plot(res_single)
```

### Leave-one-out analysis

A	plot	of	a	leave-one-out	analysis	shows	the	influence	of	individual	SNPs	on	the	overall	effect	estimate,	which	can	beused	to	identifyimportant	outliers.	

```{r, warning=FALSE}
res_loo <- mr_leaveoneout(dat)
```

```{r, warning=FALSE, message=FALSE}
mr_leaveoneout_plot(res_loo)
```

\pagebreak



# Part. Multivariable MR: Childhood BMI and Adult BMI as exposures, `r params$mediator_name` as outcome 

## Multivariable MR results, using 2SMR package (IVW)

```{r}
# put all exposure and full gwas dat into lists
exposure_list <- list(early_bmi_exp, adult_bmi_exp)
full_gwas_list<- list(early_bmi_gwas, adult_bmi_gwas)

# create exposure_dat format
exposure_dat <- get_mv_exposures(exposure_list, full_gwas_list, clump_exposures = T)
```


```{r}
#Next, also extract those SNPs from the outcome.
if (data_lookup$format == "mrbase"){
  outcome_dat <- extract_outcome_data(snps = exposure_dat$SNP, 
                                      outcomes = data_lookup$mrbase.id)
} else {
  outcome_dat <- mediator %>% filter(SNP %in% exposure_dat$SNP)
}
#Once the data has been obtained, harmonise so that all are on the same reference allele.
mvdat <- mv_harmonise_data(exposure_dat, outcome_dat)

#Finally, perform the multivariable MR analysis
res_bmis <- mv_multiple(mvdat)

mv_res_bmis<- res_bmis$result %>%
              split_outcome() %>% 
              split_exposure() %>% 
              separate(outcome, "outcome", sep="[(]") %>% 
              generate_odds_ratios()
mv_res_bmis$outcome.full<-mediator_name

report_dir <- paste0(results_path_sub, data_lookup$trait_file_name, "/", "report_",Sys.Date())
report_dir_w_backslash <- gsub(" ", "\\ ", report_dir, fixed=T) # create path vector escaping spaces, otherwise sytem call cant process it
if(!dir.exists(report_dir)){ system(paste("mkdir -p", report_dir_w_backslash))}
write_tsv(mv_res_bmis, paste0(report_dir, "/mvmr_BMIs-", data_lookup$trait_file_name, "_using_2SMR.tsv"))
```


```{r, echo = F}
mv_res_bmis<-mv_res_bmis %>% select(-id.exposure, -id.outcome, -outcome.full)
print(mv_res_bmis)
```


## Sensitivity tests, using MVMR package
```{r, echo=F, warning =F}
if (data_lookup$format == "mrbase"){
    mvmr_input <- make_mvmr_input(exposure_dat, exposure_list, outcome.id.mrbase = data_lookup$mrbase.id)
} else {
    mvmr_input <- make_mvmr_input(exposure_dat, exposure_list, outcome.data = mediator)
}
mvmr_out <- format_mvmr(BXGs = mvmr_input$XGs %>% select(contains("beta.")),
                        BYG = mvmr_input$YG$beta.outcome,
                        seBXGs = mvmr_input$XGs %>% select(contains("se.")),
                        seBYG = mvmr_input$YG$se.outcome,
                        RSID = mvmr_input$XGs$SNP)
```

_Test for weak instruments_

```{r, echo=T}
strength_mvmr(r_input=mvmr_out, gencov=0)
```

```{r, echo=T}
#
#s_yy<-matrix(c(1,0.197458,0.197458,1), nrow=2, ncol=2)
#
#colnames(s_yy) = rownames(s_yy) = c("Childhood BMI", "Adult BMI")
#
#
#se_matrix<- mvmr_input$XGs %>% select(contains("se.")) %>% as.data.frame()
#gen_cov<-MVMR::phenocov_mvmr(Pcov = s_yy,
#                       seBXGs = se_matrix )
#
#strength_mvmr(r_input=mvmr_out, gencov=gen_cov)
```


_Test for horizontal pleiotropy_
```{r, echo=T}
pleiotropy_mvmr(r_input=mvmr_out, gencov=0)
```

\pagebreak

# Part. Multivariable MR: Childhood BMI and `r params$mediator_name` as exposures, Breast cancer as outcome 


## Multivariable MR results, using 2SMR package (IVW)

```{r}
# put all exposure and full gwas dat into lists
exposure_list <- list(early_bmi_exp, exposure_mediator)
if (data_lookup$format == "mrbase"){
  mediator <- extract_outcome_data(snps = exposure_list %>% 
                                     purrr::reduce(bind_rows) %>%
                                     pull(SNP), 
                                   outcomes = data_lookup$mrbase.id)
  mediator$outcome <- mediator_name}
full_gwas_list <- list(early_bmi_gwas, mediator)

# create exposure_dat format
exposure_dat <- get_mv_exposures(exposure_list, full_gwas_list, clump_exposures = T) 
```


```{r}
#Next, also extract those SNPs from the outcome.
outcome_dat <- extract_outcome_data(exposure_dat$SNP, params$breastcancer_dataset)

#Once the data has been obtained, harmonise so that all are on the same reference allele.
mvdat <- mv_harmonise_data(exposure_dat, outcome_dat)

#Finally, perform the multivariable MR analysis
res <- mv_multiple(mvdat)

mv_res<- res$result %>%
              split_outcome() %>% 
              split_exposure() %>% 
              separate(outcome, "outcome", sep="[(]") %>% 
              generate_odds_ratios()


report_dir <- paste0(results_path_sub, data_lookup$trait_file_name, "/", "report_",Sys.Date())
report_dir_w_backslash <- gsub(" ", "\\ ", report_dir, fixed=T) # create path vector escaping spaces, otherwise sytem call cant process it
if(!dir.exists(report_dir)){ system(paste("mkdir -p", report_dir_w_backslash))}
write_tsv(mv_res, paste0(report_dir, "/mvmr_earlyBMI-",data_lookup$trait_file_name,"-BC_using_2SMR.tsv"))
```


```{r, echo = F}
mv_res<-mv_res %>% select(-id.exposure, -id.outcome)
print(mv_res)
```

## Sensitivity tests, using MVMR package
```{r, echo=F,  warning =F}
mvmr_input <- make_mvmr_input(exposure_dat, exposure_list, outcome.id.mrbase = params$breastcancer_dataset)
mvmr_out <- format_mvmr(BXGs = mvmr_input$XGs %>% select(contains("beta.")),
                        BYG = mvmr_input$YG$beta.outcome,
                        seBXGs = mvmr_input$XGs %>% select(contains("se.")),
                        seBYG = mvmr_input$YG$se.outcome,
                        RSID = mvmr_input$XGs$SNP)
```

_Test for weak instruments_

```{r, echo=T}
strength_mvmr(r_input=mvmr_out, gencov=0)
```

_Test for horizontal pleiotropy_
```{r, echo=T}
pleiotropy_mvmr(r_input=mvmr_out, gencov=0)
```


\pagebreak

# Part. Multivariable MR: Childhood BMI, Adult BMI, and `r params$mediator_name` as exposures, Breast cancer as outcome 


## Multivariable MR results, using 2SMR package (IVW)

```{r, include}
# put all exposure and full gwas dat into lists
exposure_list <- list(early_bmi_exp, adult_bmi_exp, exposure_mediator)
if (data_lookup$format == "mrbase"){
  mediator <- extract_outcome_data(snps = exposure_list %>% purrr::reduce(bind_rows)  %>% pull(SNP), 
                                      outcomes = data_lookup$mrbase.id)
  mediator$outcome <- mediator_name
}
full_gwas_list<- list(early_bmi_gwas, adult_bmi_gwas,  mediator)

# create exposure_dat format
exposure_dat <- get_mv_exposures(exposure_list, full_gwas_list, clump_exposures = T)
```


```{r}
#Next, also extract those SNPs from the outcome.
outcome_dat <- extract_outcome_data(exposure_dat$SNP, params$breastcancer_dataset)

#Once the data has been obtained, harmonise so that all are on the same reference allele.
mvdat <- mv_harmonise_data(exposure_dat, outcome_dat)

#Finally, perform the multivariable MR analysis
res <- mv_multiple(mvdat)

mv_3exp<- res$result %>%
              split_outcome() %>% 
              split_exposure() %>% 
              separate(outcome, "outcome", sep="[(]") %>% 
              generate_odds_ratios()

report_dir <- paste0(results_path_sub, data_lookup$trait_file_name, "/", "report_",Sys.Date())
report_dir_w_backslash <- gsub(" ", "\\ ", report_dir, fixed=T) # create path vector escaping spaces, otherwise sytem call cant process it
if(!dir.exists(report_dir)){ system(paste("mkdir -p", report_dir_w_backslash))}
write_tsv(mv_3exp, paste0(report_dir, "/mvmr_adultBMI-earlyBMI-", data_lookup$trait_file_name,"-BC_using_2SMR.tsv"))
```

```{r, echo = F}
mv_3exp<-mv_3exp %>% select(-id.exposure, -id.outcome)
print(mv_3exp)
```

## Sensitivity tests, using MVMR package
```{r, echo=F,  warning =F}
mvmr_input<-make_mvmr_input(exposure_dat, exposure_list, outcome.id.mrbase = params$breastcancer_dataset)
mvmr_out <- format_mvmr(BXGs = mvmr_input$XGs %>% select(contains("beta.")),
                        BYG = mvmr_input$YG$beta.outcome,
                        seBXGs = mvmr_input$XGs %>% select(contains("se.")),
                        seBYG = mvmr_input$YG$se.outcome,
                        RSID = mvmr_input$XGs$SNP)

```

_Test for weak instruments_
```{r, echo=T}
strength_mvmr(r_input=mvmr_out, gencov=0)
```

_Test for horizontal pleiotropy_
```{r, echo=T}
pleiotropy_mvmr(r_input=mvmr_out, gencov=0)
```


\pagebreak
# Summary {-}


#### Univariable MR:  BMI -> Breast cancer (total effect) {-}

```{r, echo=F}
bc_data <- stringr::str_split(params$breastcancer_dataset, "-")[[1]][3]
uni_BC<-read_tsv(paste0(results_path, "bmi/merged/merged_mr_mediators_bmi-to-BCAC_", bc_data,".tsv")) %>% 
  filter(method == "Inverse variance weighted") %>%
  select(exposure, outcome,b,lo_ci, up_ci, or, or_lci95, or_uci95, pval, se, nsnp) %>% 
  tidy_pvals()
uni_BC%>% 
  kable(.) %>%
  kable_styling()
```


#### Miltivariable MR:  BMI -> Breast cancer (direct effect) {-}

```{r, echo=F}
mv_BC <- read_tsv(paste0(results_path, "bmi/merged/mvmr_BMIs-BCAC_", bc_data,".tsv")) %>% 
  select(exposure, outcome,b,lo_ci, up_ci, or, or_lci95, or_uci95, pval, se, nsnp) %>% 
  tidy_pvals()
mv_BC %>% 
  kable(.) %>%
  kable_styling()
```




#### Univarible MR: BMI -> `r params$mediator_name` (total effect) {-}

```{r, echo=F}
uni_med<- out %>% filter(method == "Inverse variance weighted") %>%
  select(exposure, outcome,b,lo_ci, up_ci, or, or_lci95, or_uci95, pval, se, nsnp) %>% 
  tidy_pvals()
uni_med %>% 
  kable(.) %>%
  kable_styling()
```

#### Univariable MR: `r params$mediator_name` -> Breast cancer (total effect) {-}

```{r, echo=F}
uni_med_BC<-res_tidy %>%  filter(method == "Inverse variance weighted") %>%
  select(exposure, outcome,b,lo_ci, up_ci, or, or_lci95, or_uci95, pval, se, nsnp) %>% 
  tidy_pvals()
uni_med_BC %>% 
  kable(.) %>%
  kable_styling()
```


#### Multivariable MR: Childhood BMI + Adult BMI -> `r params$mediator_name` (direct effect) {-}

```{r, echo=F}
mv_med <- mv_res_bmis %>% select(exposure, outcome,b,lo_ci, up_ci, or, or_lci95, or_uci95, pval, se, nsnp) %>% 
  tidy_pvals()
mv_med %>% 
  kable(.) %>%
  kable_styling()
```


#### Multivariable MR: Early BMI + `r params$mediator_name` -> Breast cancer (direct effect) {-}

```{r, echo=F}
mv_med2_BC <-mv_res %>% select(exposure, outcome,b,lo_ci, up_ci, or, or_lci95, or_uci95, pval, se, nsnp) %>% 
  tidy_pvals()
mv_med2_BC %>% 
  kable(.) %>%
  kable_styling()
```

#### Multivariable MR: Adult BMI + Early BMI + `r params$mediator_name` -> Breast cancer (direct effect) {-}

```{r, echo=F}
mv_med3_BC <-mv_3exp %>% select(exposure, outcome,b,lo_ci, up_ci, or, or_lci95, or_uci95, pval, se, nsnp) %>% 
  tidy_pvals()
mv_med3_BC %>% 
  kable(.) %>%
  kable_styling()
```


\pagebreak


# Mediation analysis {-}

Mediation analysis can be used to dissect the total causal effect of the risk factor on the outcome into an indirect effect ofthe risk factor on the outcome via themediator, and a direct effect of the risk factor on the outcome not via the mediator.

Exploring indirect effect of mediator (`r params$mediator_name`) in `Chidhood BMI -> Breast Cancer` inverse association

## Difference method {-}

INDIRECT = TOTAL (of exposure, univ) - DIRECT (of exposure, mvmr)


```{r, echo=F, message=""}
x<- uni_BC %>% filter(exposure == "Childhood BMI") %>% pull(b)
y<- mv_med2_BC %>% filter(exposure == "Childhood BMI")  %>% pull(b)
z <- x-y
x<-round(x,3)
y<-round(y,3)
z<-round(z,3)
print(paste("INDIRECT = ", x," - ", y, " = ", z))
```

SE of INDIRECT effect (difference) = sqrt(SE TOTAL^2 + SE DIRECT^2) 

```{r, echo=F}
se.x <- uni_BC %>% filter(exposure == "Childhood BMI") %>% pull(se)
se.y <- mv_med2_BC %>% filter(exposure == "Childhood BMI")  %>% pull(se)

se.z = sqrt(se.x^2 + se.y^2)
se.x<-round(se.x,3)
se.y<-round(se.y,3)
se.z<-round(se.z,3)
print(paste("SE (INDIRECT) = sqrt(", se.x,"^2 + ", se.y, "^2) = ", se.z))
```

Estimated _direct_ effect of the mediator (`r params$mediator_name`) from MVMR model (BMI+mediator -> BC): 
```{r, echo=F}
y<- mv_med2_BC %>% filter(exposure == mediator_name)  %>% pull(b)
print(paste0( y))
```



## Product method (2 ways) {-}

INDIRECT = TOTAL (exposure -> mediator) x TOTAL (mediator -> outcome)


```{r, echo=F}
x<- uni_med %>% filter(exposure == "Childhood BMI") %>% pull(b)
y<- uni_med_BC  %>% filter(exposure == mediator_name) %>% pull(b)
z<- x * y
x<-round(x,3)
y<-round(y,3)
z<-round(z,3)
print(paste("INDIRECT = ", x," x ", y, " = ", z))
```


`OR`

INDIRECT = TOTAL (exposure -> mediator) x DIRECT (of mediator , mvmr)
```{r, echo=F}
# or
x<- uni_med %>% filter(exposure == "Childhood BMI") %>% pull(b)
y<- mv_med2_BC %>% filter(exposure == mediator_name)  %>% pull(b)
z<- x * y
x<-round(x,3)
y<-round(y,3)
z<-round(z,3)
print(paste("INDIRECT = ", x," x ", y, " = ", z))
```

Unable to calculate SE for product method using 2SMR output.

<br>
**Note on mediation analysis with binary outcomes**
from _Burgess et al 2017_

If all variables are continuous, and all relationships between variables are linear, then the total effect is equal to the direct effect plus the indirect effect.
But it is common for the outcome in a MR investigation to be a binary variable, such as disease status. In this case, typically genetic associations are obtained from logistic regression, and represent log odds ratios. Odds ratios are noncollapsible, meaning that they do not average intuitively, and they depend on the choice of covariate adjustment even in the absence ofconfounding. This means that differences between causal estimates from Difference and Product methods may arise due to _non-collapsibility_ rather than mediation. In practice,  we would recommend providing estimates of the total and direct effects, but not the indirect effect, as calculation of the indirect effect relies on the linearity of the relationships that cannot occur with a binary outcome. The total and direct effects still have interpretations as population-averaged causal effects (conditional on the mediator for the direct effect), representing the average change in the outcome resulting from intervening on the population distribution of the risk factor (while keeping the mediator constant for the direct effect). Substantial differences between these estimates would still be informative about the causal pathway from the risk factor to the outcome.



