---
title: "multivarMR"
author: "Marina Vabistsevits"
date: "10/03/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(eval=FALSE)
library(tidyr)
library(tibble)
library(dplyr)
library(readr)
library(vroom)
library(TwoSampleMR)
library(MendelianRandomization)
#install_github("WSpiller/MVMR")
library(MVMR)
```


```{r}
# set path for pre-calculated data, outside the code repo
# `local` / `remote` (reading data from RDSF)
currently_working_env = "local"
source("../set_paths.R")
set_paths(currently_working_env)

# metafile
data_lookup<-read_csv(paste0("../metadata/data_lookup.csv")) 
# load mvmr functions
source("../functions_mvmr.R")
```

# Load all BMI files
```{r, cache=TRUE, message=F}
# Load BMI exposures
early_bmi_file <- data_lookup %>% filter(trait == "Childhood BMI") %>% pull(tophits_data)
early_bmi_exp <- read_tsv(paste0(data_path_tophits, early_bmi_file))
dim(early_bmi_exp) ## 115 in new

adult_bmi_file <- data_lookup %>% filter(trait == "Adult BMI") %>% pull(tophits_data)
adult_bmi_exp <- read_tsv(paste0(data_path_tophits, adult_bmi_file))
dim(adult_bmi_exp) # 173 in new

# Load BMI outcomes
early_bmi_gwas_file <- data_lookup %>% filter(trait == "Childhood BMI") %>% pull(full_data)
early_bmi_gwas <- vroom(paste0(data_path_gwas, early_bmi_gwas_file))
dim(early_bmi_gwas) 

adult_bmi_gwas_file <- data_lookup %>% filter(trait == "Adult BMI") %>% pull(full_data)
adult_bmi_gwas <- vroom(paste0(data_path_gwas, adult_bmi_gwas_file))
dim(adult_bmi_gwas) 

```


```{r}
# load data for the third exposure
igf_gwas<- vroom(paste0(data_path_gwas, "igf_females_imputed_TIDY_OUTCOME.txt.gz"))
igf_gwas$outcome <-"IGF"
dim(igf_gwas)

igf_exp <- read_exposure_data(
          filename = paste0(data_path_tophits, "igf_females.txt"),
          sep = "\t",
          snp_col = "SNP",
          beta_col = "BETA",
          se_col = "SE",
          effect_allele_col = "ALLELE1",
          other_allele_col = "ALLELE0",
          eaf_col = "A1FREQ",
          pval_col = "P_BOLT_LMM_INF"
      )
igf_exp$exposure <- "IGF"
```


```{r}
### test on testosterone data
all.files.tsv <- list.files(path = data_path_tophits, pattern = "*tsv", full.names = T)
biotest<-vroom(all.files.tsv[3])
freetest<-vroom(all.files.tsv[9])
totaltest<-vroom(all.files.tsv[16])

biotest_gwas<- vroom(paste0(data_path_gwas, "biotest_female_invnormal_imputed_TIDY_OUTCOME.txt.gz"))
biotest_gwas$outcome <- "biotest"
freetest_gwas<- vroom(paste0(data_path_gwas, "freetest_female_invnormal_imputed_TIDY_OUTCOME.txt.gz"))
freetest_gwas$outcome <- "freetest"
totaltest_gwas<- vroom(paste0(data_path_gwas, "totaltest_female_invnormal_imputed_TIDY_OUTCOME.txt.gz"))
totaltest_gwas$outcome <- "totaltest"
```




```{r}
# testosterone with SHBG
get_names <- function(path_and_name){
    tmp<-path_and_name %>% 
      basename(.) %>% 
      strsplit(., ".", fixed = TRUE)
  return(tmp[[1]][1])
}
read_exposure_custom<-function(exposure_file){
  # assiming all provided files have these colnames
  
  out <- read_exposure_data(
        filename = exposure_file,
        sep = "\t",
        snp_col = "SNP",
        beta_col = "BETA",
        se_col = "SE",
        effect_allele_col = "ALLELE1",
        other_allele_col = "ALLELE0",
        eaf_col = "A1FREQ",
        pval_col = "P_BOLT_LMM_INF"
    )
  exposure_file <- get_names(exposure_file)
  out$exposure <- exposure_file
  return(out)
}
all.files.txt <- list.files(path = data_path_tophits, pattern = "*txt", full.names = T)

shbg_premeno <- read_exposure_custom(all.files.txt[9])
shbg_postmeno <- read_exposure_custom(all.files.txt[8])
test_premeno <- read_exposure_custom(all.files.txt[11])
test_postmeno <- read_exposure_custom(all.files.txt[10])

shbg_premeno_gwas<- vroom(paste0(data_path_gwas, "shbg_no_meno_imputed_TIDY_OUTCOME.txt.gz"))
shbg_premeno_gwas$outcome <- "shbg_premeno"
shbg_postmeno_gwas<- vroom(paste0(data_path_gwas, "shbg_meno_imputed_TIDY_OUTCOME.txt.gz"))
shbg_postmeno_gwas$outcome <- "shbg_postmeno"

test_premeno_gwas<- vroom(paste0(data_path_gwas, "test_no_meno_imputed_TIDY_OUTCOME.txt.gz"))
test_premeno_gwas$outcome <- "test_premeno"
test_postmeno_gwas<- vroom(paste0(data_path_gwas, "test_meno_imputed_TIDY_OUTCOME.txt.gz"))
test_postmeno_gwas$outcome <- "test_postmeno"

```



**TwoSampleMR multivariable MR**

## Set all inputs for multivariable MR
### NB there is a way to set all of these manually, which is a long complex progess, but if deeper understanding od what is happening is required, go to section `Manual MV`


**2SMR MV**
```{r}
### creating list of data to run

# put all exposure and full gwas dat into lists
exposure_list <- list(early_bmi_exp, adult_bmi_exp)#, igf_exp)
full_gwas_list<- list(early_bmi_gwas, adult_bmi_gwas)#, igf_gwas)

### testoterone test
exposure_list <- list( freetest, totaltest,biotest)
full_gwas_list<- list(freetest_gwas, totaltest_gwas, biotest_gwas)

### testoterone + SHBG test
exposure_list <- list( shbg_premeno, shbg_postmeno)
full_gwas_list<- list(shbg_premeno_gwas, shbg_postmeno_gwas) 
exposure_list <- list( test_premeno, test_postmeno)
full_gwas_list<- list(test_premeno_gwas, test_postmeno_gwas) 

exposure_list <- list( test_premeno, shbg_premeno)
full_gwas_list<- list(test_premeno_gwas, shbg_premeno_gwas) 
exposure_list <- list( test_postmeno, shbg_postmeno)
full_gwas_list<- list(test_postmeno_gwas, shbg_postmeno_gwas) 

###test bio, total, shbg
exposure_list <- list(totaltest,biotest, shbg_premeno)
full_gwas_list<- list( totaltest_gwas, biotest_gwas, shbg_premeno_gwas)
exposure_list <- list(totaltest,biotest, shbg_postmeno)
full_gwas_list<- list( totaltest_gwas, biotest_gwas, shbg_postmeno_gwas)



# IFG, fasting insulin
exposure_mediator1 <- extract_instruments("ieu-a-778") %>% clump_data()
exposure_mediator1$exposure <- "fasting insulin"

exposure_list <- list(igf_exp, exposure_mediator1)
  mediator <- extract_outcome_data(snps = exposure_list %>% 
                                       purrr::reduce(bind_rows) %>%
                                       pull(SNP), 
                                     outcomes = "ieu-a-778")
  mediator$outcome <- unique(exposure_mediator1$exposure)
    
full_gwas_list <- list(igf_gwas, mediator)

# IFG, fasting insulin, HOMA-IR + childhood BMI
exposure_mediator1 <- extract_instruments("ieu-a-778") %>% clump_data()
exposure_mediator1$exposure <- "fasting insulin"
exposure_mediator2 <- extract_instruments("ieu-a-771", p1=10e-7) 
exposure_mediator2$exposure <- "HOMA-IR"

exposure_list <- list(igf_exp, exposure_mediator1, exposure_mediator2, early_bmi_exp)
  mediator1 <- extract_outcome_data(snps = exposure_list %>%  purrr::reduce(bind_rows) %>% pull(SNP), 
                                     outcomes = "ieu-a-778")
  mediator1$outcome <- unique(exposure_mediator1$exposure)
  
  mediator2 <- extract_outcome_data(snps = exposure_list %>%  purrr::reduce(bind_rows) %>% pull(SNP), 
                                     outcomes = "ieu-a-771")
  mediator2$outcome <- unique(exposure_mediator2$exposure)
    
full_gwas_list <- list(igf_gwas, mediator1, mediator2, early_bmi_gwas)


# IFG, fasting insulin, + childhood BMI
exposure_mediator1 <- extract_instruments("ieu-a-778") %>% clump_data()
exposure_mediator1$exposure <- "fasting insulin"


exposure_list <- list(igf_exp, exposure_mediator1, early_bmi_exp)
  mediator1 <- extract_outcome_data(snps = exposure_list %>%  purrr::reduce(bind_rows) %>%pull(SNP), 
                                     outcomes = "ieu-a-778")
  mediator1$outcome <- unique(exposure_mediator1$exposure)
  

full_gwas_list <- list(igf_gwas, mediator1, early_bmi_gwas)

# IFG, fasting insulin, + childhood BMI + adult BMI
exposure_mediator1 <- extract_instruments("ieu-a-778") %>% clump_data()
exposure_mediator1$exposure <- "fasting insulin"


exposure_list <- list(igf_exp, exposure_mediator1, early_bmi_exp, adult_bmi_exp)
  mediator1 <- extract_outcome_data(snps = exposure_list %>% purrr::reduce(bind_rows) %>% pull(SNP), 
                                     outcomes = "ieu-a-778")
  mediator1$outcome <- unique(exposure_mediator1$exposure)
  

full_gwas_list <- list(igf_gwas, mediator1, early_bmi_gwas, adult_bmi_gwas)

```

```{r}
# create exposure_dat format
exposure_dat <- get_mv_exposures(exposure_list, full_gwas_list, clump_exposures = T)


#Next, also extract those SNPs from the outcome.
outcome_dat <- extract_outcome_data(exposure_dat$SNP, 'ieu-a-1126')
dim(outcome_dat)

#Once the data has been obtained, harmonise so that all are on the same reference allele.
mvdat <- mv_harmonise_data(exposure_dat, outcome_dat)


#Finally, perform the multivariable MR analysis
res <- mv_multiple(mvdat)


mv_res<- res$result %>%
              split_outcome() %>% 
              split_exposure() %>% 
              tidyr::separate(outcome, "outcome", sep="[(]") %>% 
              generate_odds_ratios()
mv_res
library(kableExtra)
mv_res %>%  select(-id.exposure, -id.outcome)  %>%  kable(.) %>% kable_styling()

#write_tsv(mv_res, paste0(results_path, "mvmr_BMI-to-BC_using_2SMR.tsv"))
```

Now we are going to use other packages to do multivariable MR. 
Originally, data had to be manually wrangled in a sprecial format to be accepted by these, but then I figures out I can just reshape inputs for 2SMR.
For manual wrangling see section `Manual MV`.


**MVMR (Wes's package)**
```{r}
# create MVMR package input
mvmr_input<-make_mvmr_input(exposure_dat, exposure_list, outcome.id.mrbase="ieu-a-1126")

# MVMR
mvmr_out <- format_mvmr(BXGs = mvmr_input$XGs %>% select(contains("beta.")),
                        BYG = mvmr_input$YG$beta.outcome,
                        seBXGs = mvmr_input$XGs %>% select(contains("se.")),
                        seBYG = mvmr_input$YG$se.outcome,
                        RSID = mvmr_input$XGs$SNP)

## sensitivity tests

#Test for weak instruments
sres<-strength_mvmr(r_input=mvmr_out, gencov=0)


#Test for horizontal pleiotropy
pres<-pleiotropy_mvmr(r_input=mvmr_out, gencov=0)
pres

# sensitivity test in presence of pheno matric
s_yy<-matrix(c(1,0.197458,0.197458,1), nrow=2, ncol=2)
s_yy<-cbind(c(1,0.7684388),c(0.7684388,1)) # only on instruments

colnames(s_yy) = rownames(s_yy) = c("Childhood BMI", "Adult BMI")

se_matrix<- mvmr_input$XGs %>% select(contains("se.")) %>% as.data.frame()
gen_cov<-MVMR::phenocov_mvmr(Pcov = s_yy,
                       seBXGs = se_matrix )

strength_mvmr(r_input=mvmr_out, gencov=gen_cov)
pleiotropy_mvmr(r_input=mvmr_out, gencov=gen_cov)


#Estimate causal effects
mvmr_res<-ivw_mvmr(r_input=mvmr_out)

# effect adjusted for Q het
#mvmr_res_q<-qhet_mvmr(r_input=mvmr_out, pcor=s_yy, CI=T, iterations=2) ## NOT WORKING

mvmr_res_q2<-qhet_mvmr_tmp(r_input=mvmr_out, pcor=s_yy, CI=T, iterations=2)
mvmr_res_q5<-qhet_mvmr_tmp(r_input=mvmr_out, pcor=s_yy, CI=T, iterations=5)
mvmr_res_q10<-qhet_mvmr_tmp(r_input=mvmr_out, pcor=s_yy, CI=T, iterations=10)
mvmr_res_q50<-qhet_mvmr_tmp(r_input=mvmr_out, pcor=s_yy, CI=T, iterations=50)

#mvmr_res <- mvmr(mvmr_out, 0, 1) ##OLD

res_df <- mvmr_res %>% as.data.frame() %>% 
  rownames_to_column("exposure") %>% 
  rename(b=Estimate) %>% 
  rename(se="Std. Error") %>% 
  rename(pval="Pr(>|t|)") %>% 
  select(-c(`t value`)) %>% 
  generate_odds_ratios()

res_df$exposure<-unique(exposure_dat$exposure)

write_tsv(res_df, paste0(results_path, "mvmr_BMI-to-BC_using_MVMR.tsv"))

```

**MendelianRandomization IVW**
```{r}
mr_mvivw_res <- mr_mvivw(mr_mvinput(bx = as.matrix(XGs[,c("betaX1", "betaX2")]), 
                                    bxse = as.matrix(XGs[,c("seX1", "seX2")]),
                                    by = YG$beta.outcome, 
                                    byse =YG$se.outcome, 
                                    exposure = unique(exposure_dat$exposure) ))

get_odds_ratios_mvivw<-function(mr_result){
  # create empty df
  df<-setNames(data.frame(matrix(ncol = 9, nrow = 2)),
               c("exposure", 'b', 'se',  'pval' , 'lo_ci', 'up_ci' , 'or', 'or_lci95', 'or_uci95'))
  # extract results from MR object
  df$exposure <- mr_result$Exposure
  df$b <- mr_result$Estimate
  df$se <- mr_result$StdError
  df$pval <- mr_result$Pvalue
  # calculated OR and CIs
  df$lo_ci <- df$b  - 1.96 * df$se
  df$up_ci <- df$b  + 1.96 * df$se
  df$or <- round(exp(df$b),4)
  df$or_lci95<-round(exp(df$lo_ci),4)
  df$or_uci95<-round(exp(df$up_ci),4)
  return(df)
}

mr_mvivw_res_out <- get_odds_ratios_mvivw(mr_mvivw_res)
write_tsv(mr_mvivw_res_out, paste0(results_path, "mvmr_BMI-to-BC_using_MenRandIVW.tsv"))
```

**MendelianRandomization MR-Egger**

```{r}
#MVMR-Egger
mr_mvegger_res <- mr_mvegger(mr_mvinput(bx = as.matrix(XGs[,c("betaX1", "betaX2")]), 
                                    bxse = as.matrix(XGs[,c("seX1", "seX2")]),
                                    by = YG$beta.outcome, 
                                    byse =YG$se.outcome,
                                    exposure = unique(exposure_dat$exposure) ),
                                    orientate = 1)

get_odds_ratios_mvegger<-function(mr_result){
  # create empty df
  df<-setNames(data.frame(matrix(ncol = 9, nrow = 2)),
               c("exposure", 'b', 'se',  'pval' , 'lo_ci', 'up_ci' , 'or', 'or_lci95', 'or_uci95'))
  # extract results from MR object
  df$exposure <- mr_result$Exposure
  df$b <- mr_result$Estimate
  df$se <- mr_result$StdError.Est
  df$pval <- mr_result$Pvalue.Est
  # calculated OR and CIs
  df$lo_ci <- df$b  - 1.96 * df$se
  df$up_ci <- df$b  + 1.96 * df$se
  df$or <- round(exp(df$b),4)
  df$or_lci95<-round(exp(df$lo_ci),4)
  df$or_uci95<-round(exp(df$up_ci),4)
  return(df)
}

mr_mvegger_res_out <- get_odds_ratios_mvegger(mr_mvegger_res)
write_tsv(mr_mvegger_res_out, paste0(results_path, "mvmr_BMI-to-BC_using_MenRandEgger.tsv"))

```




**End here!**


## *Manual MV*
## The long detailed way of building all exposure and outcome references manually 

```{r}
# combine exposures for later
exposures <- rbind(early_bmi_exp, adult_bmi_exp)
```


```{r}
# get early BMI exposure SNPs [exp1] in adultBMI GWAS [GWAS2]
adult_bmi_out <- adult_bmi_gwas %>% filter(SNP %in% early_bmi_exp$SNP)
harmonised_1 <- harmonise_data(early_bmi_exp, adult_bmi_out)
dim(harmonised_1)

# get adult BMI exposusure SNPs  [exp2] in earlyBMI GWAS [GWAS1]
early_bmi_out <- early_bmi_gwas %>% filter(SNP %in% adult_bmi_exp$SNP)
harmonised_2 <- harmonise_data(adult_bmi_exp, early_bmi_out)
dim(harmonised_2)
```


```{r}
# get exposure1 SNPs and exposure2 SNPs in GWAS1
# exp1 SNPs in GWAS1
exp1_gwas1 <- harmonised_1 %>% 
            select(SNP, beta.outcome, se.outcome) %>%
            rename(betaX1 = beta.outcome) %>%
            rename(seX1 = se.outcome)
# exp2 SNPs in GWAS1
exp2_gwas1 <- harmonised_2 %>%
            select(SNP, beta.exposure, se.exposure) %>%
            rename(betaX1 = beta.exposure) %>% 
            rename(seX1 = se.exposure)

dat_gwas1 <- rbind(exp1_gwas1, exp2_gwas1) %>% distinct() 
dim(dat_gwas1)

# get exposure1 SNPs and exposure2 SNPs in GWAS2
# exp2 SNPs in GWAS2
exp2_gwas2 <- harmonised_2 %>%
            select(SNP, beta.outcome, se.outcome) %>%
            rename(betaX2 = beta.outcome) %>%
            rename(seX2 = se.outcome)

# exp1 SNPs in GWAS2
exp1_gwas2 <-harmonised_1 %>%
            select(SNP, beta.exposure, se.exposure) %>%
            rename(betaX2 = beta.exposure) %>% 
            rename(seX2 = se.exposure)

dat_gwas2<-rbind(exp2_gwas2, exp1_gwas2) %>% distinct()
dim(dat_gwas2)
# join data
exposures_joined <- full_join(dat_gwas1, dat_gwas2, by="SNP") %>% distinct()
dim(exposures_joined)


### NB ####
# there are some weird almost-duplicates, but not sure what they mean;
exposures_joined %>% count(SNP) %>% arrange(desc(n)) 
almost_dups<-exposures_joined %>% count(SNP) %>% filter(n==2) %>% pull(SNP)
exposures_joined %>% filter(SNP %in% almost_dups) %>% View()
# keeping just the first instance of duplicated SNPs
exposures_joined <-exposures_joined[!duplicated(exposures_joined$SNP),] 

# check NAs
sum(is.na(exposures_joined))

# filter out SNPs that have same b+se in both exposures(thing happens when SNP was founf in both) ??
exposures_joined <- exposures_joined %>% 
  mutate(drop= ifelse(betaX1 == betaX2 & seX1 == seX2, 1, 0)) %>% 
  filter(drop == 0) %>% select(-drop) 

```

```{r}
# Outcomes

# extract SNPs for both exposures from outcome dataset 
outcome_dat <- extract_outcome_data(snps = unique(exposures_joined$SNP),
                                    outcomes ="ieu-a-1126")
# harmonize datasets 
outcome_harmonised <- harmonise_data(exposures, outcome_dat)
dim(outcome_harmonised)
```

```{r}
# Create variables for MV
# remove factors structure in SNPs, and sort by SNP (YGs and XGs must have SNPs in the same order)
YG <- outcome_harmonised %>% 
              select("SNP", "beta.outcome", "se.outcome") %>%
              distinct() %>%
              mutate(SNP = as.character(SNP)) %>%
              arrange((SNP)) 
XGs <- exposures_joined %>% 
              filter(SNP %in% outcome_harmonised$SNP) %>%
              mutate(SNP = as.character(SNP)) %>%
              arrange((SNP)) 

# some checks
stopifnot(dim(XGs)[1]==dim(YG)[1])
unique(YG$SNP %in% XGs$SNP)
unique(XGs$SNP %in% YG$SNP)
all.equal(YG$SNP, XGs$SNP)

```



```{r}
# MVMR
mvmr_out <- format_mvmr(BXGs = XGs[,c("betaX1", "betaX2")],
                        BYG = YG$beta.outcome,
                        seBXGs = XGs[,c("seX1", "seX2")],
                        seBYG = YG$se.outcome,
                        RSID = XGs$SNP)
mvmr_res <- mvmr(mvmr_out, 0, 1)

res_df <- mvmr_res$coef %>% as.data.frame() %>% 
          rownames_to_column("exposure") %>% 
          rename(b=Estimate) %>% 
          rename(se="Std. Error") %>% 
          rename(pval="Pr(>|t|)") %>% 
          generate_odds_ratios()

res_df
```


```{r}
#MVMR-IVW
mr_mvivw_res <- mr_mvivw(mr_mvinput(bx = as.matrix(XGs[,c("betaX1", "betaX2")]), 
                                bxse = as.matrix(XGs[,c("seX1", "seX2")]),
                                by = YG$beta.outcome, 
                                byse =YG$se.outcome,
                                exposure = c("Adult BMI", "Childhood BMI")))
get_odds_ratios_mvivw<-function(mr_result){
  # create empty df
  df<-setNames(data.frame(matrix(ncol = 9, nrow = 2)),
               c("exposure", 'b', 'se',  'pval' , 'lo_ci', 'up_ci' , 'or', 'or_lci95', 'or_uci95'))
  # extract results from MR object
  df$exposure <- mr_result$Exposure
  df$b <- mr_result$Estimate
  df$se <- mr_result$StdError
  df$pval <- mr_result$Pvalue
  # calculated OR and CIs
  df$lo_ci <- df$b  - 1.96 * df$se
  df$up_ci <- df$b  + 1.96 * df$se
  df$or <- round(exp(df$b),4)
  df$or_lci95<-round(exp(df$lo_ci),4)
  df$or_uci95<-round(exp(df$up_ci),4)
  return(df)
}
print(get_odds_ratios_mvivw(mr_mvivw_res))

#MVMR-Egger
mr_mvegger_res <- mr_mvegger(mr_mvinput(bx = as.matrix(XGs[,c("betaX1", "betaX2")]), 
                                    bxse = as.matrix(XGs[,c("seX1", "seX2")]),
                                    by = YG$beta.outcome, 
                                    byse =YG$se.outcome,
                                    exposure = c("Adult BMI", "Childhood BMI")),
                                    orientate = 1)
get_odds_ratios_mvegger<-function(mr_result){
  # create empty df
  df<-setNames(data.frame(matrix(ncol = 9, nrow = 2)),
               c("exposure", 'b', 'se',  'pval' , 'lo_ci', 'up_ci' , 'or', 'or_lci95', 'or_uci95'))
  # extract results from MR object
  df$exposure <- mr_result$Exposure
  df$b <- mr_result$Estimate
  df$se <- mr_result$StdError.Est
  df$pval <- mr_result$Pvalue.Est
  # calculated OR and CIs
  df$lo_ci <- df$b  - 1.96 * df$se
  df$up_ci <- df$b  + 1.96 * df$se
  df$or <- round(exp(df$b),4)
  df$or_lci95<-round(exp(df$lo_ci),4)
  df$or_uci95<-round(exp(df$up_ci),4)
  return(df)
}
print(get_odds_ratios_mvegger(mr_mvegger_res))

```










##### ignore this, keeping for ref


```{r}
### THIS WORKS

id_exposure <- c("ukb-b-3768", "ukb-b-12405")

#mv_extract_exposures :
  
	# Get best instruments for each exposure
	exposure_dat <- extract_instruments(id_exposure)
	temp <- exposure_dat
	temp$id.exposure <- 1
	temp <- clump_data(temp)
	exposure_dat <- subset(exposure_dat, SNP %in% temp$SNP)


	# Get effects of each instrument from each exposure
	d1 <- extract_outcome_data(exposure_dat$SNP, # all SNPs in both exposures, clumped
	                           id_exposure, # extract  from both full GWAS of exposures
	                           proxies=TRUE)
	stopifnot(length(unique(d1$id)) == length(unique(id_exposure)))
	d1 <- subset(d1, mr_keep.outcome)
	d2 <- subset(d1, id.outcome != id_exposure[1]) # dropped data for the first exposure -- exp2 is outcome
	d1 <- convert_outcome_to_exposure(subset(d1, id.outcome == id_exposure[1])) # covert second trait to  exposure format  -- exp1 is exposure

	# Harmonise against the first id
	d <- harmonise_data(exposure_dat = d1, 
	                    outcome_dat = d2, action=2)

	# Only keep SNPs that are present in all
	tab <- table(d$SNP)
	keepsnps <- names(tab)[tab == length(id_exposure)-1]
	d <- subset(d, SNP %in% keepsnps)
	
	# Reshape exposures
	dh1 <- subset(d,  # for first trait get exposure cols
	              id.outcome == id.outcome[1], select=c(SNP, exposure, id.exposure, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, pval.exposure))
	dh2 <- subset(d, # for the other get outcome cols
	              select=c(SNP, outcome, id.outcome, effect_allele.outcome, other_allele.outcome, eaf.outcome, beta.outcome, se.outcome, pval.outcome))
	names(dh2) <- gsub("outcome", "exposure", names(dh2)) # rename outcome to exposure in these
	
	dh <- rbind(dh1, dh2)

	
```



