---
title: "MR with mediators"
author: "Marina Vabistsevits"
date: "09/03/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(plyr)
library(tidyr)
library(purrr)
library(readr)
library(tibble)
library(dplyr)
library(data.table)
library(vroom)
library(TwoSampleMR)
library(ggplot2)
library(cowplot)
library(wesanderson)
```

```{r}
# set path for pre-calculated data, outside the code repo
# `local` / `remote` (reading data from RDSF)
currently_working_env = "local"
source("set_paths.R")

```

**Exposures**
```{r}
early_bmi_file <- paste0(data_path_tophits, "earlyBMI_UKB_tophits.csv")

# as exposure
early_bmi_exp <- read_exposure_data(
    filename = early_bmi_file,
    sep = ",",
    snp_col = "SNP",
    beta_col = "Beta",
    se_col = "Standard Error",
    effect_allele_col = "Effect allele",
    other_allele_col = "Other allele",
    eaf_col = "Effect allele frequency",
    pval_col = "P",
    gene_col = "Closest gene",
)
early_bmi_exp$exposure <- "Childhood BMI"
dim(early_bmi_exp) # 135


adult_bmi_file <- paste0(data_path_tophits, "adultBMI_UKB_tophits.csv")

# as exposure
adult_bmi_exp <- read_exposure_data(
    filename = adult_bmi_file,
    sep = ",",
    snp_col = "SNP",
    beta_col = "Beta",
    se_col = "Standard Error",
    effect_allele_col = "Effect allele",
    other_allele_col = "Other allele",
    eaf_col = "Effect allele frequency",
    pval_col = "P",
    gene_col = "Closest gene"
)
adult_bmi_exp$exposure <- "Adult BMI"
dim(adult_bmi_exp) # 215

```


# Analysis start here

As GWAS files are very large, going to read and process only one at time (also prossible may need to explore them interactively)
( NB scroll down to loop to do them all)
```{r}
# list and read files in data dir
all.files <- list.files(path = data_path_gwas, pattern = "*TIDY_OUTCOME.txt.gz", full.names = T)
all.files<-all.files[c(2,3,7,8,18,19)]
get_names_v2 <- function(path_and_name){
    tmp<-path_and_name %>% 
      basename(.) %>% 
      strsplit(., ".", fixed = TRUE) 
     tmp2 <- gsub("_invnormal_imputed_TIDY_OUTCOME", "", tmp[[1]][1])
  return(tmp2)
}
```


**Outcomes**

```{r}
# load full GWAS data and subset to exposure SNPs

# start with earlyBMI as exposure
current_file <- all.files[6]
outcome_mediator <- vroom(current_file) 
current_gwas <- get_names_v2(current_file)
outcome_mediator$outcome <- current_gwas


outcome_mediator_early <- outcome_mediator %>% filter(SNP %in% early_bmi_exp$SNP)
outcome_mediator_adult <- outcome_mediator %>% filter(SNP %in% adult_bmi_exp$SNP)

rm(outcome_mediator)
```



## Run MR on one overalping mediator 

**Harmonise**
```{r}        
dat_early <- harmonise_data(exposure_dat = early_bmi_exp, 
                        outcome_dat = outcome_mediator_early)
dat_adult <- harmonise_data(exposure_dat = adult_bmi_exp, 
                        outcome_dat = outcome_mediator_adult)
```
**Perform MR**

```{r}
res_early <- mr(dat_early) 
res_adult <- mr(dat_adult) 
```


**Tidy up results**
- Split outcome names
- Generate odds ratios with 95% confidence intervals

```{r}
res_early_tidy<- res_early %>%
              split_outcome() %>% 
              split_exposure() %>% 
              separate(outcome, "outcome", sep="[(]") %>% 
              generate_odds_ratios()

res_adult_tidy<- res_adult %>%
              split_outcome() %>% 
              split_exposure() %>%
              separate(outcome, "outcome", sep="[(]") %>% 
              generate_odds_ratios()

out <- rbind(res_early_tidy, res_adult_tidy ) %>% arrange(method) %>% select(-c("id.outcome", "id.exposure"))

write_tsv(out, paste0(results_path, "MR_BMI-to-", current_gwas,".tsv"))
```

**Sensitivity analyses**
```{r}
# Heterogenity
het_early <- mr_heterogeneity(dat_early) # Cochranâ€™s Q to detect pleiotropy in a MR context
het_adult <- mr_heterogeneity(dat_adult) 

# Scatter plot
p1_early <- mr_scatter_plot(res_early, dat_early)
p1_adult <- mr_scatter_plot(res_adult, dat_adult)

# Single SNP MR analysis + Forest plot + Funnel plot
res_single_early <- mr_singlesnp(dat_early)
p2_early <- mr_forest_plot(res_single_early)
p4_early <- mr_funnel_plot(res_single_early)

res_single_adult <- mr_singlesnp(dat_adult)
p2_adult <- mr_forest_plot(res_single_adult)
p4_adult <- mr_funnel_plot(res_single_adult)


# Leave-one-out analysis
res_loo_early <- mr_leaveoneout(dat_early)
p3_early <- mr_leaveoneout_plot(res_loo_early)

res_loo_adult <- mr_leaveoneout(dat_adult)
p3_adult <- mr_leaveoneout_plot(res_loo_adult)

```




# Creating a loop to do all of them

```{r}
all.files <- list.files(path = data_path_gwas, pattern = "*TIDY_OUTCOME.txt.gz", full.names = T)
all.files<-all.files[c(2,3,7,8,18,19)]

for (i in 1:length(all.files)){
  
    
  # load full GWAS data (outcomes) and subset to exposure SNPs
  outcome_mediator <- vroom(all.files[i]) 
  current_gwas<-get_names_v2(all.files[i])
  print(paste0("Read in ", all.files[i]))
  outcome_mediator$outcome <- current_gwas
  
  outcome_mediator_early <- outcome_mediator %>% filter(SNP %in% early_bmi_exp$SNP)
  outcome_mediator_adult <- outcome_mediator %>% filter(SNP %in% adult_bmi_exp$SNP)
  
  rm(outcome_mediator)
    
  # Harmonise 
  
  dat_early <- harmonise_data(exposure_dat = early_bmi_exp, 
                          outcome_dat = outcome_mediator_early)
  dat_adult <- harmonise_data(exposure_dat = adult_bmi_exp, 
                          outcome_dat = outcome_mediator_adult)
  
  rm(outcome_mediator_early)
  rm(outcome_mediator_adult)
  
  # Perform MR
  
  res_early <- mr(dat_early, method_list=c('mr_ivw','mr_egger_regression','mr_weighted_median', 'mr_wald_ratio')) 
  res_adult <- mr(dat_adult, method_list=c('mr_ivw','mr_egger_regression','mr_weighted_median', 'mr_wald_ratio'))   
    
  
  # Tidy up results and save
  
  res_early_tidy<- res_early %>%
                split_outcome() %>% 
                split_exposure() %>% 
                separate(outcome, "outcome", sep="[(]") %>% 
                generate_odds_ratios()
  
  res_adult_tidy<- res_adult %>%
                split_outcome() %>% 
                split_exposure() %>%
                separate(outcome, "outcome", sep="[(]") %>% 
                generate_odds_ratios()
  
  out <- rbind(res_early_tidy, res_adult_tidy ) %>% arrange(method) %>% select(-c("id.outcome", "id.exposure"))
  
  write_tsv(out, paste0(results_path, "MR_BMI-to-", current_gwas,".tsv"))
    
  print(paste0("Finished MR for ", current_gwas)) 
}

```


```{r}
# workflow if the data is in MRBase
exposures<-c("ebi-a-GCST003368", "ieu-a-1002","prot-a-1724","prot-a-2326","prot-a-1209","prot-a-1936","prot-a-2327")
exposures <-c("ukb-b-3768", "ieu-a-1095", "ukb-b-12405", "ukb-b-1209", "ukb-b-17422","ieu-a-1004")
exposures <- c("ieu-a-778", "ieu-a-775", "ieu-a-777", "ieu-a-773", "ieu-a-771", "ieu-a-770", "ieu-a-758")#, "ebi-a-GCST003367")
exposures<-c("ieu-a-775")
for (i in 1:length(exposures)){
  
    
  # load full GWAS data (outcomes) and subset to exposure SNPs
  outcome_mediator_early <- extract_outcome_data(
            snps = early_bmi_exp$SNP,
            outcome = exposures[i])
  outcome_mediator_adult <- extract_outcome_data(
            snps = adult_bmi_exp$SNP,
            outcome = exposures[i])
    
  # Harmonise 
  
  dat_early <- harmonise_data(exposure_dat = early_bmi_exp, 
                          outcome_dat = outcome_mediator_early)
  dat_adult <- harmonise_data(exposure_dat = adult_bmi_exp, 
                          outcome_dat = outcome_mediator_adult)
  
  rm(outcome_mediator_early)
  rm(outcome_mediator_adult)
  
  # Perform MR
  
  res_early <- mr(dat_early, method_list=c('mr_ivw','mr_egger_regression','mr_weighted_median', 'mr_wald_ratio')) 
  res_adult <- mr(dat_adult, method_list=c('mr_ivw','mr_egger_regression','mr_weighted_median', 'mr_wald_ratio'))   
    
  
  # Tidy up results and save
  
  res_early_tidy<- res_early %>%
                split_outcome() %>% 
                split_exposure() %>% 
                separate(outcome, "outcome", sep="[(]") %>% 
                generate_odds_ratios()
  
  res_adult_tidy<- res_adult %>%
                split_outcome() %>% 
                split_exposure() %>%
                separate(outcome, "outcome", sep="[(]") %>% 
                generate_odds_ratios()
  
  out <- rbind(res_early_tidy, res_adult_tidy ) %>% arrange(method) %>% select(-c("id.outcome", "id.exposure"))
  data_lookup<-read_csv(paste0("metadata/data_lookup.csv")) %>% filter(mrbase.id==exposures[i])
  out$outcome <- data_lookup$trait
  
  write_tsv(out, paste0(results_path, "glycemic_traits/merged/tmp/MR_BMI-to-", data_lookup$trait ,".tsv"))
    
  print(paste0("Finished MR for ", exposures[i])) 
}

```

```{r}
# load all results files into a table to view together
all.files_mr <- list.files(path = results_path, pattern = "MR_BMI*", full.names = T)
all.files_mr<- all.files_mr[c(2:7,14:18)] # BC one
all.files_mr<- all.files_mr[c(6:9,15:18)] # BC one
l <- lapply(all.files_mr, fread)

df <- l %>% reduce(rbind) %>% 
      # add nice names
         mutate(outcome=ifelse(outcome == "igf_females", "IGF", 
                        ifelse(outcome == "oest_females", "Oestradiol", 
                        ifelse(outcome == "oest_meno", "Oestradiol post-menopause", 
                        ifelse(outcome == "oest_no_meno", "Oestradiol pre-menopause", 
                        ifelse(outcome == "shbg_meno", "SHBG post-menopause",                             
                        ifelse(outcome == "shbg_no_meno", "SHBG pre-menopause",                         
                        ifelse(outcome == "test_meno", "Testosterone post-menopause",                     
                        ifelse(outcome == "test_no_meno", "Testosterone pre-menopause",
                        ifelse(outcome == "biotest_female", "Testosterone (bioavailable)",
                        ifelse(outcome == "biotest_female_nomeno", "Testosterone (bioavailable) pre-meno",
                        ifelse(outcome == "biotest_female_meno", "Testosterone (bioavailable) post-meno",
                        ifelse(outcome == "freetest_female", "Testosterone (free)",
                        ifelse(outcome == "freetest_female_nomeno", "Testosterone (free) pre-meno",
                        ifelse(outcome == "freetest_female_meno", "Testosterone (free) post-meno",
                        ifelse(outcome == "totaltest_female", "Testosterone (total)",
                        ifelse(outcome == "totaltest_female_nomeno", "Testosterone (total) pre-meno",
                        ifelse(outcome == "totaltest_female_meno", "Testosterone (total) post-meno" ,outcome            
                        )))))))))))))))))) %>% 
          # create overall hormone label
          mutate(hormone = ifelse(grepl("IGF", outcome), "IGF",
                             ifelse(grepl("Oestradiol", outcome), "Oestradiol",
                             ifelse(grepl("SHBG", outcome), "SHBG",
                             ifelse(grepl("Testosterone", outcome), "Testosterone", NA
                              )))))

# subset to each MR method and save
result_ivw <- df %>% filter(method == "Inverse variance weighted") %>% 
               select("outcome" , "exposure", "hormone","b", "lo_ci", "up_ci", "pval", "nsnp") 
write_tsv(result_ivw, paste0(results_path, "BMI-to-mediators_IVW_table.tsv"))

result_egger <- df %>% filter(method == "MR Egger") %>% 
               select("outcome" , "exposure", "hormone","b", "lo_ci", "up_ci","pval") 
write_tsv(result_egger, paste0(results_path, "BMI-to-mediators_MREgger_table.tsv"))

result_egger <- df %>% filter(method == "Weighted median") %>% 
               select("outcome" , "exposure", "hormone","b", "lo_ci", "up_ci","pval") 
write_tsv(result_egger, paste0(results_path, "BMI-to-mediators_WMedian_table.tsv"))

```


```{r}

dat<- read_tsv(paste0(results_path, "BMI-to-mediators_IVW_table.tsv"))

pal <- wes_palette("Darjeeling1")[c(2,3,4,5)]
p<-ggplot(dat, aes(y=exposure, x=b, label=outcome, shape=exposure, colour=hormone)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=.3) +
  geom_point(size=2)+##, shape=c(20,16))+#, color = "orange") +
  scale_color_manual(values=pal)+
  scale_shape_manual(values = c(19,20)) +
  #coord_fixed(ratio=.3) +
  geom_vline(xintercept=0, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~outcome, ncol=1)+
  labs(color = "",y = "", x = "Effect size",
       title="Univariate MR results, 95% CI")+
  theme(legend.position = "none")

ggsave(paste0(results_path, "Figures/BMI-to-mediators_IVW_ORplot.png"),
       plot=p, scale=1, 
       width=17, height=20,
       units=c("cm"), dpi=200, limitsize=F)

```

```{r}
# add a table of number to the plot
library(gridExtra)
library(kableExtra)

mytext<-dat %>% 
  select('exposure', 'outcome',"b", "lo_ci", "up_ci", 'pval' ,'nsnp') %>% 
  mutate(pval= as.character(pval)) %>% 
  mutate_if(is.numeric, round, digits=2) %>% 
  mutate(pval=as.numeric(pval),
  pval=scales::scientific(pval, digits = 2)) %>%  
  arrange(desc(exposure)) %>% 
  rename("Exposure" = exposure,
         "Outcome" = outcome,
         "Effect size"=b, 
         "Lower CI"= lo_ci,
         "Upper CI"= up_ci) %>% 
  arrange(Outcome) 


mytext %>% 
  kable(.) %>%
  kable_styling()

t1 <- tableGrob(mytext, theme=ttheme_minimal(), rows=NULL)
grid.arrange(p, t1, nrow=1)
```



```{r}
### special case for testosterone
dat<- read_tsv(paste0(results_path, "BMI-to-Testosterone_IVW_table.tsv"))
dat<-dat %>% 
  mutate(measure = ifelse(grepl("free", outcome), "Free", 
                   ifelse(grepl("bioavailable", outcome), "Bioavailable",
                   ifelse(grepl("total", outcome), "Total", "Total (old)")))) %>% 
  mutate(menopause = ifelse(grepl("pre", outcome), "pre-menopause", 
                   ifelse(grepl("post", outcome), "post-menopause", "all"))) %>% 
  #mutate(measure = factor(measure, levels=c("Total", "Total (old)", "Free", "Bioavailable"))) %>% 
  mutate(menopause = factor(menopause, levels=c( "pre-menopause", "post-menopause", "all"))) %>% 
  arrange(menopause)

# creatting dummy data rows fro total,old, all
tmp<-dat %>% filter(measure=="Total" & menopause == "all") %>% 
             mutate(b=NA, lo_ci=NA , up_ci=NA) %>% 
             mutate(measure = "Total (old)" )
dat <- rbind(dat, tmp)
                     

pal<-unname(yarrr::piratepal("google"))
P<-ggplot(dat, aes(y=exposure, x=b, label=outcome,  colour=measure)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=.3) +
  geom_point(size=2)+##, shape=c(20,16))+#, color = "orange") +
  scale_color_manual(values=pal)+
  #scale_shape_manual(values = c(15,19,21)) +
  #coord_fixed(ratio=.3) +
  geom_vline(xintercept=0, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~outcome, ncol=1)+
  labs(color = "", shape = "", y = "", x = "Effect size",
       title="      MR results for testosterone, 95% CI\n")+
  theme(legend.position = "top")
```


```{r}
# make faceted testosterone plot
p1<-dat %>% filter(measure == "Free") %>% 
  ggplot(aes(y=exposure, x=b, label=measure, colour=measure)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=.3) +
  geom_point(size=2)+##, shape=c(20,16))+#, color = "orange") +
  scale_color_manual(values=pal[1])+
  #scale_shape_manual(values = c(15,19,21)) +
  xlim(-0.068, 0.62)+
  geom_vline(xintercept=0, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~menopause, ncol=3)+
  labs(color = "",y = "", x = "",
       title="                     MR results for testosterone, 95% CI\n")+
  theme(legend.position = "right")
p2<-dat %>% filter(measure == "Bioavailable") %>% 
  ggplot(aes(y=exposure, x=b, label=measure,  colour=measure)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=.3) +
  geom_point(size=2)+##, shape=c(20,16))+#, color = "orange") +
  scale_color_manual(values=pal[2])+
  #scale_shape_manual(values = c(15,19,21)) +
  xlim(-0.068, 0.62)+
  geom_vline(xintercept=0, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~menopause, ncol=3)+
  labs(color = "",y = "", x = "",
       title="")+
  theme(legend.position = "right")
p3<-dat %>% filter(measure == "Total") %>% 
  ggplot(aes(y=exposure, x=b, label=measure,  colour=measure)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=.3) +
  geom_point(size=2)+##, shape=c(20,16))+#, color = "orange") +
  scale_color_manual(values=pal[3])+
  #scale_shape_manual(values = c(15,19,21)) +
  xlim(-0.068, 0.62)+
  geom_vline(xintercept=0, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~menopause, ncol=3)+
  labs(color = "",y = "", x = "",
       title="")+
  theme(legend.position = "right")
p4<-dat %>% filter(measure == "Total (old)") %>% 
  ggplot(aes(y=exposure, x=b, label=measure,  colour=measure)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=.3) +
  geom_point(size=2)+##, shape=c(20,16))+#, color = "orange") +
  scale_color_manual(values=pal[4])+
  #scale_shape_manual(values = c(15,19,21)) +
  xlim(-0.068, 0.62)+
  geom_vline(xintercept=0, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~menopause, ncol=3)+
  labs(color = "",y = "", x = "Effect size",
       title="")+
  theme(legend.position = "right")
library(patchwork)
p<- ( p1 ) / (p2) / (p3) / (p4)

```

## Reproductive traits

```{r}
# load all results files into a table to view together
all.files_mr <- list.files(path = paste0(results_path, "reproductive_traits/"), pattern = "MR_BMI*", full.names = T)
all.files_mr<- all.files_mr[c(1:5, 14)] # BC one
l <- lapply(all.files_mr, fread)

df <- l %>% reduce(rbind) %>% 
          # create overall hormone label
          mutate(trait_type = ifelse(grepl("menarche", outcome), "menarche",
                             ifelse(grepl("menopause", outcome), "menopause",
                             ifelse(grepl("live birth", outcome), "live birth",
                             ifelse(grepl("births", outcome), "births", NA
                              )))))

# subset to each MR method and save
result_ivw <- df %>% filter(method == "Inverse variance weighted") %>% 
               select("outcome" , "exposure", "trait_type","b", "lo_ci", "up_ci", "pval", "nsnp") 
write_tsv(result_ivw, paste0(results_path, "reproductive_traits/", "BMI-to-reproductive_traits_IVW_table.tsv"))

result_egger <- df %>% filter(method == "MR Egger") %>% 
               select("outcome" , "exposure", "trait_type","b", "lo_ci", "up_ci","pval") 
write_tsv(result_egger, paste0(results_path, "reproductive_traits/", "BMI-to-reproductive_traits_MREgger_table.tsv"))

result_egger <- df %>% filter(method == "Weighted median") %>% 
               select("outcome" , "exposure", "trait_type","b", "lo_ci", "up_ci","pval") 
write_tsv(result_egger, paste0(results_path, "reproductive_traits/", "BMI-to-reproductive_traits_WMedian_table.tsv"))

```

```{r}

dat<- read_tsv(paste0(results_path, "reproductive_traits/" ,"BMI-to-reproductive_traits_IVW_table.tsv"))

pal<-rev(unname(yarrr::piratepal("info2")))
p<-ggplot(dat, aes(y=exposure, x=b, label=outcome, shape=exposure, colour=trait_type)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=.3) +
  geom_point(size=2)+##, shape=c(20,16))+#, color = "orange") +
  scale_color_manual(values=pal)+
  scale_shape_manual(values = c(19,20)) +
  #coord_fixed(ratio=.3) +
  geom_vline(xintercept=0, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~outcome, ncol=1)+
  labs(color = "",y = "", x = "Effect size",
       title="Univariate MR results, 95% CI")+
  theme(legend.position = "none")

ggsave(paste0(results_path, "Figures/BMI-to-mediators_IVW_ORplot.png"),
       plot=p, scale=1, 
       width=17, height=20,
       units=c("cm"), dpi=200, limitsize=F)

```

## Glycemic traits

```{r}
# load all results files into a table to view together
all.files_mr <- list.files(path = paste0(results_path,"glycemic_traits/merged/tmp/"), pattern = "MR_BMI*", full.names = T)
l <- lapply(all.files_mr, fread)

df <- l %>% reduce(rbind) %>% 
          # create overall hormone label
          mutate(trait_type = ifelse(grepl("insulin", outcome), "insulin",
                             ifelse(grepl("glucose", outcome), "glucose",
                             ifelse(grepl("HOMA-B", outcome), "HOMA-B",
                             ifelse(grepl("HOMA-IR", outcome), "HOMA-IR",
                             ifelse(grepl("HBa1c", outcome), "HBa1c", NA
                              ))))))

# subset to each MR method and save
result_ivw <- df %>% filter(method == "Inverse variance weighted") %>% 
               select("outcome" , "exposure", "trait_type","b", "lo_ci", "up_ci", "pval", "nsnp") 
write_tsv(result_ivw, paste0(paste0(results_path,"glycemic_traits/merged/"), "BMI-to-glycemic_traits_IVW_table.tsv"))

result_egger <- df %>% filter(method == "MR Egger") %>% 
               select("outcome" , "exposure", "trait_type","b", "lo_ci", "up_ci","pval") 
write_tsv(result_egger, paste0(paste0(results_path,"glycemic_traits/merged/"), "BMI-to-glycemic_traits_MREgger_table.tsv"))

result_egger <- df %>% filter(method == "Weighted median") %>% 
               select("outcome" , "exposure", "trait_type","b", "lo_ci", "up_ci","pval") 
write_tsv(result_egger, paste0(paste0(results_path,"glycemic_traits/merged/"), "BMI-to-glycemic_traits_WMedian_table.tsv"))

```

```{r}

dat<- read_tsv(paste0(paste0(results_path,"glycemic_traits/merged/"), "BMI-to-glycemic_traits_IVW_table.tsv"))

pal<-c(unname(yarrr::piratepal("pony")))
p<-ggplot(dat, aes(y=exposure, x=b, label=outcome, shape=exposure, colour=trait_type)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=.3) +
  geom_point(size=2)+##, shape=c(20,16))+#, color = "orange") +
  scale_color_manual(values=pal)+
  scale_shape_manual(values = c(19,20)) +
  #coord_fixed(ratio=.3) +
  geom_vline(xintercept=0, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~outcome, ncol=1)+
  labs(color = "",y = "", x = "Effect size",
       title="Univariate MR results, 95% CI")+
  theme(legend.position = "none")



```


```{r}
kable_it<-function(df){
  library(kableExtra)
  df %>% 
    mutate(pval= as.character(pval)) %>% 
  mutate_if(is.numeric, round, digits=2) %>% 
  mutate(pval=as.numeric(pval),
         pval=scales::scientific(pval, digits = 2)) %>% 
  kable(.) %>%
  kable_styling()
}

dat %>% kable_it()


```

