---
title: "MR_mediators_to_BC"
author: "Marina Vabistsevits"
date: "09/03/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(plyr)
library(tidyr)
library(purrr)
library(readr)
library(dplyr)
library(data.table)
library(TwoSampleMR)
library(ggplot2)
library(cowplot)
library(wesanderson)
```

```{r}
# load studies from MR base
#ao <- available_outcomes()

# set path for pre-calculated data, outside the code repo

## remote data paths (RDSF)
data_path_tophits <- "../../data/GWAS_tophits/"  # or clumped_instruments
data_path_gwas <- "../../data/GWAS_results_new/" # or simply GWAS_results 
results_path <-  "../../results/"

## local data paths
local_path<-"/Users/ny19205/OneDrive - University of Bristol/Documents - OneDrive/Mini-project2/Data/"
data_path_tophits <- paste0(local_path, "GWAS_tophits/") 
data_path_gwas <-  paste0(local_path,"GWAS_results/TIDY/") 
results_path <-   paste0(local_path, "../Results/")
```

**Exposures**
```{r}
# functions
get_names <- function(path_and_name){
    tmp<-path_and_name %>% 
      basename(.) %>% 
      strsplit(., ".", fixed = TRUE)
  return(tmp[[1]][1])
}

get_author<-function(filename){
  tmp<-filename %>% 
      strsplit(., "_", fixed = TRUE)
  return(tmp[[1]][3])
}
read_exposure_custom<-function(exposure_file){
  # assiming all provided files have these colnames
  
  out <- read_exposure_data(
        filename = exposure_file,
        sep = "\t",
        snp_col = "SNP",
        beta_col = "BETA",
        se_col = "SE",
        effect_allele_col = "ALLELE1",
        other_allele_col = "ALLELE0",
        eaf_col = "A1FREQ",
        pval_col = "P_BOLT_LMM_INF"
    )
  exposure_file <- get_names(exposure_file)
  out$exposure <- exposure_file
  return(out)
}
read_exposure_custom_v2<-function(exposure_file){
  # assiming all provided files have these colnames
  
  out <- read_exposure_data(
        filename = exposure_file,
        sep = ",",
        snp_col = "SNP",
        beta_col = "Effect",
        se_col = "SE",
        effect_allele_col = "Effect_Allele",
        other_allele_col = "Other_Allele",
        eaf_col = "Effect_Freq",
        pval_col = "P_BOLT_LMM"
    )
  exposure_file <- get_names(exposure_file)
  out$exposure <- exposure_file
  out<-clump_data(out)
  
  return(out)
}
```

```{r}
# list and read files in data dir
all.files.tsv <- list.files(path = data_path_tophits, pattern = "*tsv", full.names = T)
all.files.txt <- list.files(path = data_path_tophits, pattern = "*txt", full.names = T)
all.files.csv <- list.files(path = data_path_tophits, pattern = "*csv", full.names = T)


# read .txt files
all.files.txt<-all.files.txt[c(10:11)]
l1 <- lapply(all.files.txt, read_exposure_custom)

#read .tsv (already right columns) files
all.files.tsv<-all.files.tsv[c(1:3, 7:9, 14:16)]
l2 <- lapply(all.files.tsv, fread)

#read .tsv (already right columns) files
all.files.csv<-all.files.csv[c(2,5,6)]
l3 <- lapply(all.files.csv, read_exposure_custom_v2)


l<- append(l1,l2)
all.names <- unlist(lapply(c(all.files.txt,all.files.tsv), get_names))

# update list names to file name
for (i in 1:length(l) ){
  #print(all.files[i])
  names(l)[i] <- all.names[i]
  l[[i]]$id.exposure <- l[[i]]$exposure
  #l[[i]]$exposure <- paste( l[[i]]$exposure, get_author(all.names[i]))
}
```

Or load instruments from MR-Base
```{r}
exposures<-c("ebi-a-GCST003368", "ieu-a-1002","prot-a-1724","prot-a-2326","prot-a-1209","prot-a-1936","prot-a-2327")
exposures <-c("ukb-b-3768", "ieu-a-1095", "ukb-b-12405", "ukb-b-1209", "ukb-b-17422","ieu-a-1004")
exposures <- c("ieu-a-778", "ieu-a-775", "ieu-a-777", "ieu-a-773",  "ieu-a-770", "ieu-a-758", "ieu-a-771")#, "ebi-a-GCST003367")

# for ieu-a-771 manully extarcted instrument at 10e-7

l <- list()
for (i in 1:length(exposures)){
  l[[i]]<-extract_instruments(exposures[i])#, p1 = 10e-07)
  data_lookup<-read_csv(paste0("metadata/data_lookup.csv")) %>% filter(mrbase.id==exposures[i])
  l[[i]]$exposure <- data_lookup$trait
  print(paste0("done ",exposures[i]))
}
```



**Outcome**

BCAC, N= 228951, females only , Breast cancer (Combined Oncoarray; iCOGS; GWAS meta analysis)

```{r}
# load studies from MR base

extract_outcome_data_custom <- function(exposure_dat){
  out <- extract_outcome_data(
            snps = exposure_dat$SNP,
            outcome = 'ieu-a-1126',
            proxies = TRUE,
            rsq = 0.8, maf_threshold = 0.3) 
  return(out)
}

outcomes_l <- lapply(l, extract_outcome_data_custom)


for (i in 1:length(l)){
  l[[i]]
  print(names(l)[i])
  print(paste0("SNPs in exposure: ", dim(l[[i]])[1]))
  print(paste0("SNPs matches outcome: ", dim(outcomes_l[[i]])[1]))
}
```

```{r}
# quick venn for testosterone
library(eulerr)
s4 <- list(free = outcomes_l[["freetest_female"]]$SNP,
           bio = outcomes_l[["biotest_female"]]$SNP,
           total = outcomes_l[["totaltest_female"]]$SNP)
plot(euler(s4, shape = "ellipse"), quantities = TRUE)

s5 <- list(free = outcomes_l[["freetest_female"]]$SNP,
           bio = outcomes_l[["biotest_female"]]$SNP,
           total = outcomes_l[["totaltest_female"]]$SNP,
           premeno = outcomes_l[["test_no_meno"]]$SNP,
           postmeno = outcomes_l[["test_meno"]]$SNP)
plot(euler(s5, shape = "ellipse"), quantities = TRUE)

# with data from Ruth2020
s6<- list(bio = outcomes_l[["biotest_female"]]$SNP,
           total = outcomes_l[["totaltest_female"]]$SNP,
           free = outcomes_l[["freetest_female"]]$SNP,
           bioRuth2020 = outcomes_l[["biotest_female_Ruth2020"]]$SNP,
           totalRuth2020 = outcomes_l[["totaltest_female_Ruth2020"]]$SNP)
plot(euler(s6, shape = "ellipse"), quantities = TRUE)

s7<- list(#free = outcomes_l[["freetest_female"]]$SNP,
          # bio = outcomes_l[["biotest_female"]]$SNP,
           #total = outcomes_l[["totaltest_female"]]$SNP,
           premeno = outcomes_l[["test_no_meno"]]$SNP,
           postmeno = outcomes_l[["test_meno"]]$SNP,
           bioRuth2020 = outcomes_l[["biotest_female_Ruth2020"]]$SNP,
           totalRuth2020 = outcomes_l[["totaltest_female_Ruth2020"]]$SNP)
plot(euler(s7, shape = "ellipse"), quantities = TRUE)

# subtypes of testosterone (copy/replace each measure)
s1<-list(total_all = outcomes_l[["totaltest_female"]]$SNP,
        total_post_meno = outcomes_l[["totaltest_female_meno"]]$SNP,
        total_pre_meno = outcomes_l[["totaltest_female_nomeno"]]$SNP)
plot(euler(s1, shape = "ellipse"), quantities = TRUE)


s2<-list(total_all = outcomes_l[["totaltest_female"]]$SNP,
        total_post_meno = outcomes_l[["totaltest_female_meno"]]$SNP,
        total_pre_meno = outcomes_l[["totaltest_female_nomeno"]]$SNP,
        total_post_meno_OLD = outcomes_l[["test_meno"]]$SNP,
        total_pre_meno_OLD = outcomes_l[["test_no_meno"]]$SNP)
plot(euler(s2, shape = "ellipse"), quantities = TRUE)

# SHBG

s8<- list(shbg_meno = outcomes_l[["shbg_meno"]]$SNP,
           shbg_no_meno = outcomes_l[["shbg_no_meno"]]$SNP,
           shbg_adj_Ruth2020 = outcomes_l[["shbg_adj_Ruth2020"]]$SNP)
plot(euler(s8, shape = "ellipse"), quantities = TRUE)


```



**Harmonise**

```{r}
# harmonise each exposure to the matched outcome SNPs
harmonise_l <- list()

for (i in 1:length(l)){
   harmonise_l[[i]] <- harmonise_data(exposure_dat = l[[i]], 
                                      outcome_dat = outcomes_l[[i]])
}
```



**Perform MR**

```{r}
mr_l <- list()
for (i in 1:length(harmonise_l)){
  mr_l[[i]] <- mr(harmonise_l[[i]], method_list=c('mr_ivw','mr_egger_regression','mr_weighted_median', 'mr_wald_ratio')) 
}
```


**Tidy up results**
- Split outcome names
- Generate odds ratios with 95% confidence intervals

```{r}

tidy_result_l <- list()

for (i in 1:length(mr_l)){
  tidy_result_l[[i]] <- mr_l[[i]] %>%
                        split_outcome() %>% 
                        split_exposure() %>% 
                        separate(outcome, "outcome", sep="[(]") %>% 
                        generate_odds_ratios()
}

df <- tidy_result_l %>%
      reduce(rbind) %>% 
      arrange(method) %>% 
      select(-c("id.exposure"))

# save full data from MR analyses
#write_tsv(df, paste0(results_path, "MR_TestosteroneALL-to-BC_allMR.tsv"))
#write_tsv(df, paste0(results_path, "MR_ReproductiveTraits-to-BC_allMR.tsv"))
write_tsv(df, paste0(results_path,"glycemic_traits/merged/", "MR_GlycemicTraits-to-BC_allMR.tsv"))
```


# Hormones
```{r}

# reload data for viz/exploration
df <- read_tsv(paste0(results_path, "MR_Mediators-to-BC_allMR.tsv"))

df <- df %>% 
      # add nice names
         mutate(exposure=ifelse(exposure == "igf_females", "IGF", 
                        ifelse(exposure == "oest_females", "Oestradiol", 
                        ifelse(exposure == "oest_meno", "Oestradiol post-menopause", 
                        ifelse(exposure == "oest_no_meno", "Oestradiol pre-menopause", 
                        ifelse(exposure == "shbg_meno", "SHBG post-menopause",                        
                        ifelse(exposure == "shbg_no_meno", "SHBG pre-menopause",   
                        ifelse(exposure == "SHBG_female_inst", "SHBG",
                        ifelse(exposure == "Testosterone_female_inst", "Testosterone",  
                        ifelse(exposure == "test_meno", "Testosterone post-menopause",    
                        ifelse(exposure == "test_no_meno", "Testosterone pre-menopause",
                        ifelse(exposure == "biotest_female", "Testosterone (bioavailable)",
                        ifelse(exposure == "freetest_female", "Testosterone (free)",
                        ifelse(exposure == "totaltest_female", "Testosterone (total)", exposure     
                        )))))))))))))) %>% 
          # create overall hormone label
          mutate(hormone = ifelse(grepl("IGF", exposure), "IGF",
                             ifelse(grepl("Oestradiol", exposure), "Oestradiol",
                             ifelse(grepl("SHBG", exposure), "SHBG",
                             ifelse(grepl("Testosterone", exposure), "Testosterone", NA
                              )))))


result_ivw <- df %>% filter(method %in% c("Inverse variance weighted", "Wald ratio")) %>% 
               select("outcome" , "exposure", "hormone","or", 'method', "or_lci95", "or_uci95","b","pval", "nsnp") 

              # leveling dat does not help to reoder the plot D:
              #mutate(exposure = factor(exposure, levels=c("Oestradiol", "Oestradiol pre-menopause", "IGF", "SHBG", 
              #"SHBG pre-menopause", "SHBG post-menopause", "Testosterone", "Testosterone pre-menopause","Testosterone post-menopause")))

               
write_tsv(result_ivw, paste0(results_path, "MR_Mediators-to-BC_IVW_table.tsv"))
```


# Hormones subset

```{r}

result_ivw<-result_ivw %>% filter(!exposure %in% c("biotest_female", "freetest_female", "totaltest_female", "Testosterone", "SHBG"))

pal <- wes_palette("Darjeeling1")[c(2,3,4,5)]
p<-ggplot(result_ivw, aes(y=exposure, x=or, label=outcome, colour=hormone, shape=method)) +
  geom_errorbarh(aes(xmin=or_lci95, xmax=or_uci95), height=.3) +
  geom_point(size=2)+##, shape=c(20,16))+#, color = "orange") +
  scale_color_manual(values=pal)+
  scale_shape_manual(values = c(16,1)) +
  geom_vline(xintercept=1, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~outcome, ncol=1)+
  labs(color = "", shape ="", y = "", x = "Odds ratio",
       title="                            Univariate MR results, 95% CI")+
  theme(legend.position = "bottom",
         plot.title.position = "plot")+
  guides(colour=FALSE)

ggsave(paste0(results_path, "Mediators-to-BC_IVW_ORplot.png"),
       plot=p, scale=1, 
       width=17, height=20,
       units=c("cm"), dpi=200, limitsize=F)

```


```{r}
# make plot for non-hormonal mediators

result_ivw <- df %>% filter(method %in% c("Inverse variance weighted", "Wald ratio")) %>% 
               select("outcome" , "exposure", "id.exposure","or", 'method', "or_lci95", "or_uci95","b","pval", 'nsnp') 

pal <- wes_palette("Darjeeling1")[c(2,3,4,5)]#[c(3,2,4,5)]
p<-ggplot(result_ivw, aes(y=exposure, x=or, label=outcome, colour=id.exposure, shape=method)) +
  geom_errorbarh(aes(xmin=or_lci95, xmax=or_uci95), height=.3) +
  geom_point(size=2)+##, shape=c(20,16))+#, color = "orange") +
  scale_shape_manual(values = c(16,1)) +
  scale_color_manual(values=pal)+
  geom_vline(xintercept=1, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~outcome, ncol=1)+
  labs(color = "", shape ="", y = "", x = "Odds ratio",
       title="                            Exposures and Outcomes (MR IVW), 95% CI")+
  theme(legend.position = "bottom",
         plot.title.position = "plot")+
  guides(colour=FALSE)


```


## Testosterone
```{r}
# reload data for viz/exploration
df <- read_tsv(paste0(results_path, "MR_TestosteroneALL-to-BC_allMR.tsv"))

df <- df %>% 
      # add nice names
         mutate(exposure=ifelse(exposure == "test_meno", "Testosterone (total, old) post-menopause",                     
                        ifelse(exposure == "test_no_meno", "Testosterone (total, old) pre-menopause",
                        ifelse(exposure == "biotest_female", "Testosterone (bioavailable)",
                        ifelse(exposure == "biotest_female_nomeno", "Testosterone (bioavailable) pre-meno",
                        ifelse(exposure == "biotest_female_meno", "Testosterone (bioavailable) post-meno",
                        ifelse(exposure == "freetest_female", "Testosterone (free)",
                        ifelse(exposure == "freetest_female_nomeno", "Testosterone (free) pre-meno",
                        ifelse(exposure == "freetest_female_meno", "Testosterone (free) post-meno",
                        ifelse(exposure == "totaltest_female", "Testosterone (total)",
                        ifelse(exposure == "totaltest_female_nomeno", "Testosterone (total) pre-meno",
                        ifelse(exposure == "totaltest_female_meno", "Testosterone (total) post-meno" ,exposure       
                        )))))))))))) %>% 
        arrange(exposure) %>% 
        mutate(measure = ifelse(grepl("free", exposure), "Free", 
                   ifelse(grepl("bioavailable", exposure), "Bioavailable",
                   ifelse(grepl("total)", exposure), "Total", "Total (old)")))) %>% 
        mutate(menopause = ifelse(grepl("pre", exposure), "pre-menopause", 
                   ifelse(grepl("post", exposure), "post-menopause", "all"))) 


result_ivw <- df %>% filter(method %in% c("Inverse variance weighted", "Wald ratio")) %>% 
               select("outcome" , "exposure","or", 'method', "or_lci95", "or_uci95","b","pval", "nsnp", "measure", "menopause") 

              # leveling dat does not help to reoder the plot D:
              #mutate(exposure = factor(exposure, levels=c("Oestradiol", "Oestradiol pre-menopause", "IGF", "SHBG", 
              #"SHBG pre-menopause", "SHBG post-menopause", "Testosterone", "Testosterone pre-menopause","Testosterone post-menopause")))

               
write_tsv(result_ivw, paste0(results_path, "MR_TestosteroneALL-to-BC_IVW_table.tsv"))
```

```{r}
# testosterone
pal<-unname(yarrr::piratepal("google"))
result_ivw$exposure <- forcats::fct_inorder(result_ivw$exposure) %>% forcats::fct_rev() 
p<-ggplot(result_ivw, aes(y=exposure, x=or, label=outcome, colour= measure)) +
  geom_errorbarh(aes(xmin=or_lci95, xmax=or_uci95), height=.3) +
  geom_point(size=2)+##, shape=c(20,16))+#, color = "orange") +
  scale_color_manual(values=pal)+
  #scale_shape_manual(values = c(16,1)) +
  geom_vline(xintercept=1, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~outcome, ncol=1)+
  labs(color = "", shape ="", y = "", x = "Odds ratio",
       title="                            Univariate MR results, 95% CI")+
  theme(legend.position = "bottom",
         plot.title.position = "plot")+
  guides(colour=FALSE)


```


## Reproductive traits

```{r}
# reload data for viz/exploration
df <- read_tsv(paste0(results_path, "MR_ReproductiveTraits-to-BC_allMR.tsv"))

df <- df %>% mutate(trait_type = ifelse(grepl("menarche", exposure), "menarche",
                             ifelse(grepl("menopause", exposure), "menopause",
                             ifelse(grepl("live birth", exposure), "live birth",
                             ifelse(grepl("births", exposure), "births", NA
                              )))))
result_ivw <- df %>% filter(method %in% c("Inverse variance weighted", "Wald ratio")) %>% 
               select("outcome" , "exposure","or", 'method', "or_lci95", "or_uci95","b","pval", "nsnp", "trait_type") %>% arrange(exposure)
write_tsv(result_ivw, paste0(results_path, "MR_ReproductiveTraits-to-BC_IVW_table.tsv"))

pal<-rev(unname(yarrr::piratepal("info2")))
result_ivw$exposure <- forcats::fct_inorder(result_ivw$exposure) %>% forcats::fct_rev() 
p<-ggplot(result_ivw, aes(y=exposure, x=or, label=outcome, colour= trait_type)) +
  geom_errorbarh(aes(xmin=or_lci95, xmax=or_uci95), height=.3) +
  geom_point(size=2)+##, shape=c(20,16))+#, color = "orange") +
  scale_color_manual(values=pal)+
  #scale_shape_manual(values = c(16,1)) +
  geom_vline(xintercept=1, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~outcome, ncol=1)+
  labs(color = "", shape ="", y = "", x = "Odds ratio",
       title="                            Univariate MR results, 95% CI")+
  theme(legend.position = "bottom",
         plot.title.position = "plot")+
  guides(colour=FALSE)


```

## Glycemic traits

```{r}
# reload data for viz/exploration
df <- read_tsv(paste0(results_path,"glycemic_traits/merged/", "MR_GlycemicTraits-to-BC_allMR.tsv"))

df <- df %>% mutate(trait_type = ifelse(grepl("insulin", exposure), "insulin",
                             ifelse(grepl("glucose", exposure), "glucose",
                             ifelse(grepl("HOMA-B", exposure), "HOMA-B",
                             ifelse(grepl("HOMA-IR", exposure), "HOMA-IR",
                             ifelse(grepl("HBa1c", exposure), "HBa1c", NA
                              ))))))
result_ivw <- df %>% filter(method %in% c("Inverse variance weighted", "Wald ratio")) %>% 
               select("outcome" , "exposure","or", 'method', "or_lci95", "or_uci95","b","pval", "nsnp", "trait_type") %>% arrange(exposure)
write_tsv(result_ivw, paste0(results_path,"glycemic_traits/merged/", "MR_GlycemicTraits-to-BC_IVW_table.tsv"))

pal<-c(unname(yarrr::piratepal("pony")))
result_ivw$exposure <- forcats::fct_inorder(result_ivw$exposure) %>% forcats::fct_rev() 
p<-ggplot(result_ivw, aes(y=exposure, x=or, label=outcome, colour= trait_type)) +
  geom_errorbarh(aes(xmin=or_lci95, xmax=or_uci95), height=.3) +
  geom_point(size=2)+##, shape=c(20,16))+#, color = "orange") +
  scale_color_manual(values=pal)+
  #scale_shape_manual(values = c(16,1)) +
  geom_vline(xintercept=1, linetype='longdash') +
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~outcome, ncol=1)+
  labs(color = "", shape ="", y = "", x = "Odds ratio",
       title="                            Univariate MR results, 95% CI")+
  theme(legend.position = "bottom",
         plot.title.position = "plot")+
  guides(colour=FALSE)


```

```{r}
kable_it<-function(df){
  library(kableExtra)
  df %>% 
    mutate(pval= as.character(pval)) %>% 
  mutate_if(is.numeric, round, digits=2) %>% 
  mutate(pval=as.numeric(pval),
         pval=scales::scientific(pval, digits = 2)) %>% 
  kable(.) %>%
  kable_styling()
}

result_ivw %>% kable_it()


```


```{r}
# add a table of number to the plot
library(gridExtra)
library(kableExtra)

mytext<-result_ivw %>% 
  select('exposure', 'or',"or_lci95", "or_uci95", 'pval' ,'nsnp') %>% 
  mutate(pval= as.character(pval)) %>% 
  mutate_if(is.numeric, round, digits=2) %>% 
  mutate(pval=as.numeric(pval),
  pval=scales::scientific(pval, digits = 2)) %>%    arrange(desc(exposure)) %>% 
  rename("Exposure data" = exposure,
         OR=or, 
         "Lower CI"= or_lci95,
         "Upper CI"= or_uci95) %>% 
  separate("Exposure data", into = c('Phenotype', 'Author'), sep = '\\s(?=\\S*?$)', convert = TRUE)

mytext %>% 
  kable(.) %>%
  kable_styling()


t1 <- tableGrob(mytext, theme=ttheme_minimal(), rows=NULL)
grid.arrange(p, t1, ncol=1)
```

