---
title: "phenotypic correlation matrix"
author: "Marina Vabistsevits"
date: "01/07/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(readr)
library(vroom)
library(tidyr)
library(tibble)
library(dplyr)

#install.packages("BiocManager")
#BiocManager::install("metaCCA")
library(metaCCA)
```

```{r}
# set path for pre-calculated data, outside the code repo
# `local` / `remote` (reading data from RDSF)
currently_working_env = "local"
source("../set_paths.R")
set_paths(currently_working_env)

# metadata file that is itiratelely read in and updated
data_lookup<-paste0("../metadata/data_lookup.csv")

```



```{r}

read_for_metaCCA <- function(file_gwas, data_name) {
  # for reading in data from raw GWAS files
  
  #create custom beat and se colnames for traits
  beta = paste0(gsub("_", "-", data_name), "_b")
  se = paste0(gsub("_", "-", data_name), "_se")

  tmp <- vroom(file_gwas, col_select = c("SNP","ALLELE0", "ALLELE1", "BETA","SE")) %>% 
            rename(!!beta := "BETA",
                   !!se := "SE") %>% 
             # create `SNP` variable in all dfs
            unite(SNP, c("SNP","ALLELE0", "ALLELE1"), sep="_" ,remove=T) 
  return(tmp)
}

read_for_metaCCA_vcf <- function(file_gwas, data_name, mrbase.id) {
  # for reading in data from text files converted from GWAS VCF files
  # parse_gwas_vcf.sh  - use this script to convert GWAS VCF file to format required for this function
  
  tmp <- vroom(file_gwas) %>% 
            select(-c(contains("pval")))  %>% 
            # in positions with no rsid, in this data  chr_pos -> chr:pos to match GWAS format
            mutate(rsid = sub("_", ":", rsid)) %>% 
            # create `SNP` variable in all dfs  
            # NB seems that in VCF data SNPs are wrong way around: effect allele is ALT allele
            unite(SNP, c("rsid", "allele_1", "allele_0"), sep="_" ,remove=T) %>% 
            # switch sign at beta, because we are swapping the alleles
            mutate_at(vars(contains('_b')), list(~ .*-1)  )

  
    #create custom beta and se colnames for traits
  old_colnames<-colnames(tmp)
  colnames(tmp) <- paste0(gsub(mrbase.id, gsub("_", "-", data_name), old_colnames))
  
  return(tmp)
}

```

```{r, cache=TRUE, message=F}
# Load BMI GWAS

bmi_list <- list()

early_bmi_gwas_file <- read_csv(data_lookup)  %>% filter(trait == "Childhood BMI") %>% pull(original_file)
early_bmi_file_gwas <- paste0(data_path_gwas_raw, early_bmi_gwas_file)
early_bmi_data_name <- paste0(read_csv(data_lookup) %>% filter(trait == "Childhood BMI") %>% pull(trait_file_name)) 
bmi_list[[1]] <- read_for_metaCCA(early_bmi_file_gwas, early_bmi_data_name)

adult_bmi_gwas_file <- read_csv(data_lookup)  %>% filter(trait == "Adult BMI") %>% pull(original_file)
adult_bmi_file_gwas <- paste0(data_path_gwas_raw, adult_bmi_gwas_file)
adult_bmi_data_name <- paste0(read_csv(data_lookup) %>% filter(trait == "Adult BMI") %>% pull(trait_file_name)) 
bmi_list[[2]] <- read_for_metaCCA(adult_bmi_file_gwas, adult_bmi_data_name)

childhood_height_gwas_file <- read_csv(data_lookup)  %>% filter(trait == "Childhood height") %>% pull(original_file)
childhood_height_file_gwas <- paste0(data_path_gwas_raw, childhood_height_gwas_file)
childhood_height_data_name <- paste0(read_csv(data_lookup) %>% filter(trait == "Childhood height") %>% pull(trait_file_name)) 
bmi_list[[3]] <- read_for_metaCCA(childhood_height_file_gwas, childhood_height_data_name)


```

```{r message=F}
# specify the source of full summary stats GWAS file
data_source <- "ieu_gwas_pipeline" # ieu_gwas_pipeline : subsets to UKB data processed manually
traits <- read_csv(data_lookup) %>% filter(source == data_source) %>% pull(trait)

data_source  <-"ieu_ukb_vcf_convert" # this is just a variable for next steps
traits <- read_csv(data_lookup) %>% filter(grepl("menarche|menopause", trait)) %>% filter(!is.na(mrbase.id)) %>% pull(mrbase.id)

```


```{r message=F}

results <- data.frame()

# copy  of BMI list
l <- bmi_list

#traits<-traits[c(1,2)] # bmis only
#traits<-traits[c(2,3)] # specify traits
traits <- c("ieu-a-1004" )
print(traits)

for (i in 1:length(traits)) {
  
  n=length(l) # depends if we only added bmis or height too

  if (data_source == "ieu_gwas_pipeline"){
    current_trait<-traits[i]
    gwas_filename<- read_csv(data_lookup) %>% filter(trait == current_trait) %>% pull(original_file)
    file_gwas <- paste0(data_path_gwas_raw, gwas_filename)
    data_name <- paste0(read_csv(data_lookup) %>% filter(trait == current_trait) %>% pull(trait_file_name)) 
  
    print(paste0("Processing: ", current_trait, ", ", gwas_filename, " as ", data_name))
    
    l[[n+1]] <- read_for_metaCCA(file_gwas, data_name)
  
  } else if (data_source == "ieu_ukb_vcf_convert"){
    
    file_gwas <- paste0(data_path_gwas_raw, "from_vcf/", traits[i], ".tsv")
    data_name <- read_csv(data_lookup) %>% filter(mrbase.id == traits[i]) %>% pull(trait_file_name)
    current_trait <- read_csv(data_lookup) %>% filter(mrbase.id == traits[i]) %>% pull(trait)
    
    print(paste0("Processing: ", current_trait, ", ", file_gwas, " as ", data_name))

    l[[n+1]] <- read_for_metaCCA_vcf(file_gwas, data_name, traits[i])
  }
     

  dat <- l %>% 
    # join into a table by SNP column
    purrr::reduce(inner_join, by = "SNP") %>%  # inner join is used because raw GWAS and vcf GWAS are different
    # split into 3 cols
    separate(SNP, into=c('rsid', 'allele_0', 'allele_1'), sep="_") %>% 
    # drop indels
    filter_at(vars(starts_with("allele")), all_vars(. %in% c("A","T","G","C")) )  
  
  dim(dat)


  if (sum(is.na(dat)) !=0 ){
      print("      ~ WARNING: filtered df contains NAs ~      ")
  }
  
  print("Merging and filtering datasets ...")
  s_xy<- dat %>%  
      # join rsid+alleles to create unique rsid
      unite("rsid_w_alleles", rsid:allele_1, sep="_", remove=FALSE) %>% 
      # covert to factors (required for later)
      mutate_at(vars(starts_with("allele")), factor) %>% 
      # set rsid column as rownames
      select(-rsid) %>% column_to_rownames('rsid_w_alleles')   
  
  dim(s_xy)
    
  print("~~ Estimating phenotypic correlation matrix ~~")
  s_yy = estimateSyy( S_XY = s_xy )
  dim(s_yy)    

  
  print("Merging new rows")
  
  data_name <- gsub("_", "-", data_name) 
  tmp<- s_yy %>% as.data.frame() %>% 
    rownames_to_column('n') %>% 
    select(n, data_name) %>% 
    column_to_rownames('n') %>%  t()
  
  results <- rbind(results, tmp) %>%
            drop_na()
  
  print(results)
  print("Done")
  
   # restart
  rm(l)
  l <- bmi_list
  #print(length(l))
  
}

write_tsv(results, paste0(results_path, "pheno_correlations_tmp.tsv"))
# then same this filein metadata/



#tibble::tribble(
#                        ~mediator,   ~early_bmi,   ~adult_bmi, ~child_height,
#                            "IGF", -0.036185937,  -0.12031469,    0.02945991,
#    "Testosterone (bioavailable)", -0.003834008,  0.258677907,  -0.007487593,
#            "Testosterone (free)",  0.001187721,  0.274948957,  -0.005593064,
#                           "SHBG", -0.006185706, -0.391566531,   0.005745311,
#           "Testosterone (total)",   -0.0030827,  0.084337783,  -0.001596138,
#               "Childhood height",  0.047850338,  0.000311649,             1,
#                  "Childhood BMI",            1,  0.197457971,   0.047850338,
#                      "Adult BMI",  0.197457971,            1,   0.000311649,
#          "Age at menarche (UKB)", -0.180386304, -0.150829774,            NA,
#  "Age at first live birth (UKB)", -0.001871126, -0.160617665,            NA,
#         "Number of births (UKB)", -0.029300556,  0.047101152,            NA,
#         "Age at menopause (UKB)",  0.014678657, -0.020963617,            NA,
#                     "Chronotype",  0.016089757, -0.030266736,            NA,
#        "Age at menarche (Perry)",   -0.1035335,  -0.09516159,   -0.06769585,
#         "Age at menopause (Day)", -0.003869682,  -0.03127742,   -0.00885289
#  )



```

